---
title: "Novo Nordisk Applied Single Cell Analysis course October 2024 - scRNAseq analysis"
output:
  html_document: default
  pdf_document: default
date: "2024-10-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---

## [0.] Prerequisites

## [0.0.] Avoid gcc compiler problems
```{r}
base::Sys.setenv(
  PATH = base::paste(
      "/opt/software/gcc/11.2.0/bin",
      base::Sys.getenv("PATH"),
      sep = ":"
    )
  )
base::Sys.setenv(
    CXX = "/opt/software/gcc/11.2.0/bin/g++"
  )
```

## [0.1.] Install necessary packages
```{r eval=FALSE}
utils::install.packages("tidyverse")
utils::install.packages("here")
utils::install.packages("assertthat")

utils::install.packages("ape") # used for BuildClusterTree in [3.5.]
utils::install.packages("tidygraph")
utils::install.packages("ggraph")
utils::install.packages("clustree") # used in [3.6.]
utils::install.packages("WriteXLS")
utils::install.packages("ellipsis")
utils::install.packages("pkgbuild")
utils::install.packages("pkgload")
utils::install.packages("profvis")
utils::install.packages("remotes")
utils::install.packages("sessioninfo")
utils::install.packages("urlchecker")
utils::install.packages("usethis")
devtools::install_github("wjawaid/enrichR")
utils::install.packages("enrichR") # used in [5.9.]
```

### [0.2.] Check version of packages
#### Worfklow relies on Seurat v5 (v5.1.0).
```{r eval=FALSE}
utils::packageVersion("Seurat")
```

### [0.3.] Load necessary packages
```{r message=FALSE}
base::library("Seurat")
base::library("tidyverse")
```

---

## [1.] Read in & preprocess data

### [1.0.] Check that directory exists
```{r}
dir <- base::paste0(
    here::here("input_data/"),
    "raw_feature_bc_matrix/"
  )
base::dir.exists(paths = dir)
```

### [1.1.] Read in data as sparse matrices
#### Cellranger provides 'barcodes', 'features', 'matrix' output files
#### Data read in as dgCMatrices.
```{r}
pbmc_data <- Seurat::Read10X(
    data.dir = dir,
    gene.column = 2, # column specifying gene names in features.tsv.gz
    cell.colum = 1 # column specifying the cell ID in barcodes.tsv.gz
  )
```

### [1.2.] Inspect data & calculate sparsity
#### See https://slowkow.com/notes/sparse-matrix/ for more info on how to access data in dgCMatrices.
#### Sparse data most optimally saved in dgCMatrices.
#### Dimensions: 36601 x 733612, Sparsity: 99.92142
```{r}
# utils::str(pbmc_data)
pbmc_data$`Gene Expression`[1:5,1:5]
base::dim(pbmc_data$`Gene Expression`)
100 - (base::length(pbmc_data$`Gene Expression`@x) / base::prod(pbmc_data$`Gene Expression`@Dim))*100 # Sparsity
```

### [1.3.] Create a Seurat object from the scRNAseq data
#### Seurat objects are S3 object-oriented objects, see https://adv-r.hadley.nz/s3.html.
```{r}
pbmc_rna <- Seurat::CreateSeuratObject(
    counts = pbmc_data$`Gene Expression`, ## Specify scRNAseq
    assay = "RNA",
    project = "pbmc",
    min.cells = 3, # minimal amount of cells expressing a feature
    min.features = 200 # minimal amount of features detected per cell
  )
```

#### Alternatively create a Seurat object from both matrices.
```{r eval=FALSE}
pbmc_multi <- Seurat::CreateSeuratObject(
    counts = pbmc_data,
    assay = "Multi",
    project = "pbmc",
    min.cells = 3, # minimal amount of cells expressing a feature
    min.features = 200 # minimal amount of features detected per cell
  )
utils::str(pbmc_multi)
```

### [1.4.] Inspect the Seurat object
#### Filtering in [1.3.] reduced the number of rows and columns.
#### Dimensions: 26315 x 10987, Sparsity: 94.39873
```{r eval=FALSE}
utils::str(pbmc_rna)
base::dim(pbmc_rna) # base::dim(pbmc_rna@assays$RNA@layers$counts)
pbmc_rna@assays$RNA@layers$counts[1:5,1:5]
100 - (base::length(pbmc_rna@assays$RNA@layers$counts@x) / base::prod(pbmc_rna@assays$RNA@layers$counts@Dim))*100 # Sparsity
```

### [1.5.] Detect & count mitochondrial gene contribution
#### On terminal:  gunzip -c features.tsv.gz | cut -f2 | grep '^MT-' | head
```{r}
# base::dimnames(pbmc_rna@assays$RNA@features@.Data)[[1]]
# base::colnames(pbmc_rna)
feat_names <- base::rownames(pbmc_rna) 
feat_names[base::grep("^MT-",feat_names)]
  
pbmc_rna[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_rna,
    pattern = "^MT-"
  )
```

### [1.6.] Inspect & visualize features, including mitochondrial gene contribution
```{r}
# pbmc_rna@meta.data # Inspect new variable
# base::colnames(pbmc_rna[[]])

Seurat::VlnPlot(
    object = pbmc_rna,
    features = base::colnames(pbmc_rna[[]])[2:4],
    # layer = "counts",
    ncol = 3
  )
base::mean(pbmc_rna@meta.data$percent.mt)

Seurat::FeatureScatter(
      object = pbmc_rna,
      feature1 = "nCount_RNA",
      feature2 = "percent.mt"
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = base::c(5,10),
      linetype = "dashed"
    )

Seurat::FeatureScatter(
      object = pbmc_rna,
      feature1 = "nCount_RNA",
      feature2 = "nFeature_RNA"
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = base::c(200L,2500L),
      linetype = "dashed"
    )
```

### [1.7.] Filter cells
#### Based on inspection in [1.6.].
#### Removes empty droplets, droplets with â‰¥ 2 cells, apoptotic cells or cells with insufficiently captured transcripts.
#### Dimensions: 26315 x 7573, Sparsity: 94.87731
```{r}
pbmc_rna <- base::subset(
    x = pbmc_rna,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10
  )
# base::dim(pbmc_rna)
# 100 - (base::length(pbmc_rna@assays$RNA@layers$counts@x) / base::prod(pbmc_rna@assays$RNA@layers$counts@Dim))*100 # Sparsity
```

---

## [2.] Process data - Normalize, scale, reduce dimensions

### [2.0.] Normalization
```{r}
pbmc_rna <- Seurat::NormalizeData(
    object = pbmc_rna,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
```

### [2.1.] Inspect normalized data
```{r eval=FALSE}
utils::str(pbmc_rna)
pbmc_rna@assays$RNA@layers$data
pbmc_rna[["RNA"]]$data
```

---

#### See [5.10.] in '20241017_scRNAseq_analysis_advanced.Rmd'
#### for alternative approach using Seurat::SCTransform().

---

### [2.2.] Find most highly variably expressed genes
#### Identification of features (genes) with highest variability across all cells to apply variance-stabilizing data transformation.
```{r}
pbmc_rna <- Seurat::FindVariableFeatures(
    object = pbmc_rna,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
```

### [2.3.] Inspect variable genes
```{r}
# utils::str(pbmc_rna)
# pbmc_rna@assays$RNA@meta.data
  
lbls <- Seurat::VariableFeatures(object = pbmc_rna)[1:10]
raw <- Seurat::VariableFeaturePlot(
    object = pbmc_rna
  )
Seurat::LabelPoints(
      plot = raw,
      points = lbls,
      repel = TRUE,
      xnudge = 0,
      ynudge = 0
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = 1.25,
      linetype = "dashed"
    )
```

### [2.4.] Scale across dataset
#### Linear transformation for each feature (gene) across the variable genes to prepare principal component analysis (PCA).
#### Scale across all genes as default is to perform scaling only on the variable features.
```{r}
pbmc_rna <- Seurat::ScaleData(
    object = pbmc_rna,
    features = base::rownames(pbmc_rna),
    verbose = FALSE
  )
```

#### Alternative is to use scaling for removal of potential sources of variation, e.g., mitochondrial reads or cell cycle phase.
```{r eval=FALSE}
pbmc_rna <- Seurat::ScaleData(
    object = pbmc_rna,
    vars.to.regress = "percent.mt"
  )
```

### [2.5.] Run principal component analysis
#### RunPCA when called on a Seurat objects operates on the scale.data slot of the assay by default.
#### Seurat:::PrepDR5() function prepares internally the data to be used for PCA.
```{r}
# Seurat::DefaultAssay(object = pbmc_rna)
# SeuratObject::Layers(object = pbmc_rna)
# SeuratObject::LayerData(object = pbmc_rna, layer = "scale.data")[1:10,1:10]

pbmc_rna <- Seurat::RunPCA(
    object = pbmc_rna,
    features = Seurat::VariableFeatures(object = pbmc_rna),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
```

### [2.6.] Inspect PCA results
```{r}
# utils::str(pbmc_rna)
# pbmc_rna[["pca"]]
# pbmc_rna@reductions$pca@feature.loadings

base::print(
    pbmc_rna[["pca"]],
    dims = 1:3,
    nfeatures = 10
  )
Seurat::VizDimLoadings(
    object = pbmc_rna,
    dims = 1:3,
    reduction = "pca",
    ncol = 3,
    nfeatures = 10
  )
Seurat::DimPlot(
      object = pbmc_rna,
      reduction = "pca",
      dims = 1:2
    ) +
  theme_bw() +
  theme(
      legend.position = "none",
      aspect.ratio = 1
    )
```

### [2.7.] Inspect dimensionality of the data
```{r}
Seurat::DimHeatmap(
    object = pbmc_rna,
    dims = 1:15,
    cells = 500, # top 500 cells
    balanced = TRUE,
    nfeatures = 10,
    slot = "scale.data"
  )

Seurat::ElbowPlot(
      object = pbmc_rna,
      ndims = 50,
      reduction = "pca"
    ) +
  geom_hline(
      yintercept = 1.3,
      linetype = "dashed"
    ) +
  geom_vline(
      xintercept = base::c(5,10),
      linetype = "dashed"
    )
```

### [2.8.] Resampling method using JackStraw.
#### Not used or recommended as it is a time-consuming approach.
```{r eval = FALSE}
pbmc_rna_ <- Seurat::JackStraw(
    object = pbmc_rna,
    dims = 50,
    num.replicate = 100
  )
pbmc_rna_ <- Seurat::ScoreJackStraw(
    object = pbmc_rna_,
    dims = 1:50,
    score.thresh = 1e-05
  )
Seurat::JackStrawPlot(
    object = pbmc_rna_,
    dims = 1:50
  )
```

---

## [3.] Cell clustering
#### Seurat performs clustering based on PCA scores of features (genes) in cells.
#### Choice of included PCAs based on resampling test and Elbow plots.

### [3.0.] Calculate nearest neighbors and corresponding graph
#### k-NN graph calculation based on cell distances in PCA latent space.
#### Refinement of graph edges based on similarity between neighborhoods.
#### Uses RcppAnnoy::AnnoyEuclidean and RANN::nn2 (NNHelper) for calculating.
#### The shared nearest neighbour graph SNN_SmallestNonzero_Dist imnplemented via Rcpp (ComputeSNN).
```{r}
# Seurat::DefaultAssay(object = pbmc_rna) # assay: "RNA"
# SeuratObject::Embeddings(object = pbmc_rna[["pca"]]) # data.use: cell x PCA

pbmc_rna <- Seurat::FindNeighbors(
    object = pbmc_rna,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
```

### [3.1.] Inspect graph
```{r eval=FALSE}
utils::str(pbmc_rna)
pbmc_rna@graphs
```

### [3.2.] Find clusters
#### Louvain or SLM algorithms to iteratively form groups of cells.
#### Granularity of result adjusting the amount of identified clusters.
```{r}
pbmc_rna <- Seurat::FindClusters(
    object = pbmc_rna,
    resolution = 0.3, # <---- granularity parameter
    verbose = FALSE,
    random.seed = 0
  )
```

### [3.3.] Perform non-linear dimension reduction
```{r warning=FALSE}
pbmc_rna <- Seurat::RunUMAP(
    object = pbmc_rna,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE,
    seed.use = 42
  )
```

### [3.4.] Visualize UMAP dimension reduction
#### UMAP preserves the local distances in favor of global relationships.
```{r}
Seurat::DimPlot(
      object = pbmc_rna,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
    aspect.ratio = 1
  )
```

### [3.5.] Calculate cluster dendogram
```{r}
pbmc_rna_ <- pbmc_rna
pbmc_rna_ <- Seurat::BuildClusterTree(
    object = pbmc_rna,
    assay = "RNA",
    features = base::rownames(pbmc_rna),
    dims = 1:10,
    reduction = "pca",
    slot = "scale.data",
    verbose = FALSE
  )
# utils::str(pbmc_rna_)

# replaces Seurat::PlotClusterTree
data.tree <- SeuratObject::Tool(
    object = pbmc_rna_,
    slot = "Seurat::BuildClusterTree"
  )
ape::plot.phylo(x = data.tree, direction = "downwards")
ape::nodelabels()
```

### [3.6.] Evaluate cluster number as a function of cluster granularity / resolution
#### Incompatibilities between ggplot2 & clustree likely renders this chunk unfunctional.
```{r eval=FALSE}
pbmc_rna_ <- pbmc_rna
for(r in base::seq(0.1,1.2,0.1)){
  pbmc_rna_ <- Seurat::FindClusters(
      object = pbmc_rna_,
      resolution = r,  # <---- granularity parameter
      verbose = FALSE
    )
  }
pbmc_rna_@meta.data

clustree::clustree(
    x = pbmc_rna_[[]],
    prefix = "RNA_snn_res."
  )
```

## [3.7.] Inspect various cluster definition granularities
```{r eval=FALSE}
Seurat::DimPlot(
      object = pbmc_rna_,
      reduction = "umap",
      group.by = "RNA_snn_res.1.2" # <---- granularity parameter
    ) +
  theme_bw() +
  theme(
    aspect.ratio = 1
  )
```

---

### [3.8.] Save & load data Seurat object 'pbmc_rna'
```{r eval=FALSE}
base::saveRDS(
    object = pbmc_rna,
    file = base::paste0(
        here::here("output_data/"),
        base::format(base::Sys.time(),"%Y%m%d"),
        "_pbmc_RNA",
        ".rds"
      ),
    compress = TRUE
  )
```

```{r eval=FALSE}
pbmc_rna <- base::list.files(
      path = here::here("output_data"),
      full.names = TRUE
    ) %>%
  stringr::str_subset("RNA.rds") %>%
  base::readRDS(
      file = .
    )
```
