---
title: "Novo Nordisk Applied Single Cell Analysis Course October 2024 - scATAC-seq analysis"
output:
  html_document: default
  pdf_document: default
date: "`r Sys.Date()`"
---

###### Modified from Signac Vignettes:
###### Analyzing PBMC scATAC-seq: https://stuartlab.org/signac/articles/pbmc_vignette

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## [0.] Prerequisites
### [0.1.] Load necessary packages
```{r message=FALSE, warning=FALSE}
library(Signac) # scATAC-seq analysis tools
library(Seurat) # scRNA-seq analysis tools
library(GenomicRanges) # Process ATAC peak regions
library(TFBSTools) # Transcription factor binding site analysis
library(ggplot2) # Visualization tools
library(patchwork) # combined multiple plots

library(EnsDb.Hsapiens.v86) # Human gene annotation database
library(BSgenome.Hsapiens.UCSC.hg38) # Human genome sequence database
library(JASPAR2020) # Transcription factor motif database
```


## [1.] Import data and create seurat/signac object
### [1.1.] Specify input files paths
#### Here, we import Cell Ranger ARC pre-processed results of PBMC multiome data.
#### For other 10x scATAC or multiome datasets, you can find these files under
#### outs/ directory for each sample.
#### See https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/output/overview (scATAC)
#### or https://www.10xgenomics.com/support/software/cell-ranger-arc/latest/analysis/outputs/understanding-output (multiome)
#### for more information of 10x Genomics Cell Ranger output
```{r}
count_matrix_dir <- paste0(here::here("input_data"),"/filtered_feature_bc_matrix/")
fragments_path <- paste0(here::here("input_data"),"/10k_PBMC_Multiome_nextgem_Chromium_X_atac_fragments.tsv.gz")
metadata_path <- paste0(here::here("input_data"),"/10k_PBMC_Multiome_nextgem_Chromium_X_per_barcode_metrics.csv")
```

### [1.2.] Import count matrix
```{r message=FALSE, warning=FALSE}
pbmc_count_data <- Read10X(count_matrix_dir)
# Inspect count data
pbmc_count_data$Peaks[1:5,1:5]
```

### [1.3.] Import metadata
```{r message=FALSE, warning=FALSE}
pbmc_metadata <- read.csv(
  file = metadata_path,
  header = TRUE,
  row.names = 1
)
# Inspect metadata
pbmc_metadata[1:5,]
```

### [1.4.] Create seurat/signac object
```{r warning=FALSE}
# Create chromatinAssay object
chrom_assay <- CreateChromatinAssay(
  counts = pbmc_count_data$Peaks, ## Specify scATAC-seq
  sep = c(":", "-"),
  fragments = fragments_path,
  min.cells = 10, # include peaks detected in at least 10 cells
  min.features = 200 # include cells detecting more than 200 peaks
)

# Create seurat object from chromatinAssay object and integrate metadata
pbmc_atac <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks", # assay name
  meta.data = pbmc_metadata
)

# Remove peaks in non-standard chromosomes
peaks.keep <- seqnames(granges(pbmc_atac)) %in% standardChromosomes(granges(pbmc_atac))
pbmc_atac <- pbmc_atac[as.vector(peaks.keep), ]

# Add gene annotation information
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevels(annotation) <- paste0('chr', seqlevels(annotation))
Annotation(pbmc_atac) <- annotation
```

### [1.5.] Data structure of seurat/signac object
#### See https://stuartlab.org/signac/articles/data_structures for more 
#### information about data structures and object interaction
```{r}
# Show peak and cell number
pbmc_atac

# Show peak x cell count matrix
pbmc_atac@assays$peaks$counts[1:5,1:5]

# Show peak information
granges(pbmc_atac)

# Show gene annotation
Annotation(pbmc_atac)
```


## [2.] Quality control (QC)
#### We follow signac pipeline to use following QC metrics:
#### i. Nucleosome signal: 
####    ratio of mononucleosomal to nucleosome-free fragments
#### ii. Transcriptional start site (TSS) enrichment score: 
####     ratio of fragments centered at the TSS to fragments in TSS-flanking 
####     regions (see https://www.encodeproject.org/data-standards/terms/)
#### iii. Total number of fragments in peaks
#### iv. Fraction of fragments in peaks
#### v. Ratio reads in genomic blacklist regions
### [2.1.] Compute QC metrics
```{r}
# Compute nucleosome signal score per cell (It may take ~10 mins)
pbmc_atac <- NucleosomeSignal(object = pbmc_atac)

# Compute TSS enrichment score per cell (It may take ~10 mins)
pbmc_atac <- TSSEnrichment(object = pbmc_atac) 

# Add fraction of reads in peaks
pbmc_atac$pct_reads_in_peaks <- pbmc_atac$atac_peak_region_fragments / pbmc_atac$atac_fragments * 100

# Add blacklist ratio
pbmc_atac$blacklist_ratio <- FractionCountsInRegion(
  object = pbmc_atac, 
  assay = 'peaks',
  regions = blacklist_hg38_unified
)
```


### [2.2.] Determine threshold of QC metrics
#### The exact QC thresholds need to be adjusted based on your dataset. 
#### Consider plotting the distribution of each QC metric to accurately 
#### determine the appropriate thresholds. The primary objective is to 
#### exclude outliers while retaining the majority of cells.
```{r}
# Plot distribution of nucleosome signal
hist(pbmc_atac$nucleosome_signal,
     breaks = 20,
     main = "nucleosome_signal")
# Indicate threshold
abline(v = 1.5, # threshold
       col="red")

# Plot distribution of TSS enrichment score
hist(pbmc_atac$TSS.enrichment,
     main = "TSS.enrichment")
# Indicate threshold
abline(v = 4, # threshold
       col="red")

# Plot distribution of total number of fragments in peaks
hist(pbmc_atac$nCount_peaks,
     main = "nCount_peaks")
# Indicate threshold
abline(v = 10000, # threshold
       col="red")
abline(v = 100000, # threshold
       col="red")

# Plot distribution of fraction of fragments in peaks
hist(pbmc_atac$pct_reads_in_peaks,
     main = "pct_reads_in_peaks")
# Indicate threshold
abline(v = 40, # threshold
       col="red")

# Plot distribution of ratio reads in genomic blacklist regions
hist(pbmc_atac$blacklist_ratio,
     main = "blacklist_ratio")
# Indicate threshold
abline(v = 0.01, # threshold
       col="red")
```

### [2.3.] Filter high quality cells
#### Filter cells according to thresholds determined above
```{r}
pbmc_atac <- subset(
  x = pbmc_atac,
  subset = nucleosome_signal < 1.5 &
    TSS.enrichment > 4 &
    nCount_peaks > 10000 &
    nCount_peaks < 100000 &
    pct_reads_in_peaks > 40 &
    blacklist_ratio < 0.01 
)
```



## [3.] Dimensional reduction and clustering
### [3.1.]  Normalization and linear dimensional reduction
```{r}
# Perform term frequency-inverse document frequency (TF-IDF) normalization
pbmc_atac <- RunTFIDF(pbmc_atac)

# Feature selection
pbmc_atac <- FindTopFeatures(pbmc_atac, 
                             min.cutoff = 'q0') # Use all peaks here

# Perform singular value decomposition (SVD) for linear dimensional reduction
pbmc_atac <- RunSVD(pbmc_atac)

# Assess correlation between each LSI component and sequencing depth 
DepthCor(pbmc_atac, n = 30) 
# The first LSI component captures sequencing depth (technical variation)
# rather than biological variation.
```

### [3.2.] Non-linear dimension reduction with UMAP
```{r message=FALSE, warning=FALSE}
pbmc_atac <- RunUMAP(object = pbmc_atac, 
                     reduction = 'lsi', 
                     dims = 2:30) # Exclude the first LSI component 
```

### [3.3.] Clustering
```{r}
# Calculate nearest neighbors
pbmc_atac <- FindNeighbors(object = pbmc_atac, 
                           reduction = 'lsi', 
                           dims = 2:30) # Exclude the first LSI component

# Louvain clustering
pbmc_atac <- FindClusters(object = pbmc_atac,
                          resolution = 0.8) # higher resolution => more clusters

# Show clustering results on UMAP
DimPlot(object = pbmc_atac, 
        label = TRUE) + 
  NoLegend()
```

### [3.4.] Save seurat object
```{r}
base::saveRDS(
    object = pbmc_atac,
    file = paste0(here::here("output_data"),
                            "/pbmc_ATAC.rds")
  )
```