---
title: "Novo Nordisk Applied Single Cell Analysis Course October 2024 - scATAC-seq analysis advanced"
output:
  html_document: default
  pdf_document: default
date: "`r Sys.Date()`"
---

###### Modified from Signac Vignettes:
###### Analyzing PBMC scATAC-seq: https://stuartlab.org/signac/articles/pbmc_vignette
###### Motif analysis with Signac: https://stuartlab.org/signac/articles/motif_vignette

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## [0.] Prerequisites
### [0.1.] Load necessary packages
```{r message=FALSE, warning=FALSE}
library(Signac) # scATAC-seq analysis tools
library(Seurat) # scRNA-seq analysis tools
library(GenomicRanges) # Process ATAC peak regions
library(TFBSTools) # Transcription factor binding site analysis
library(ggplot2) # Visualization tools
library(patchwork) # Combined multiple plots

library(EnsDb.Hsapiens.v86) # Human gene annotation database
library(BSgenome.Hsapiens.UCSC.hg38) # Human genome sequence database
library(JASPAR2020) # Transcription factor motif database
```


### [1.6.] Import previous scATAC seurat object
```{r message=FALSE, warning=FALSE}
pbmc_atac <- readRDS(paste0(here::here("output_data"),
                            "/pbmc_ATAC.rds"))
```


## [4.] Cell type annotation
### [4.1.] Quantify gene activity
#### Quantify the activity of each gene by summing the fragments intersecting 
#### the gene body and promoter (TSS 2 kb upstream region)
```{r warning=FALSE}
# Generate a gene activity matrix  (It may take ~8 mins)
gene.activities <- GeneActivity(pbmc_atac)

# Add the gene activity matrix to the Seurat object as a new assay
pbmc_atac[['RNA']] <- CreateAssayObject(counts = gene.activities)

# Normalize gene activity matrix
pbmc_atac <- NormalizeData(
  object = pbmc_atac,
  assay = 'RNA',
  normalization.method = 'LogNormalize',
  scale.factor = median(pbmc_atac$nCount_RNA)
)

# Show gene activity of marker genes on UMAP
DefaultAssay(pbmc_atac) <- 'RNA'
FeaturePlot(
  object = pbmc_atac,
  features = c('MS4A1', 'CD3D', 'LEF1', 'NKG7', 'TREM1', 'LYZ'),
  pt.size = 0.1,
  max.cutoff = 'q95',
  ncol = 2
)
```


### [4.2.] Transfer cell-type labels from scRNA-seq data
```{r warning=FALSE}
# Import scRNA-seq seurat object  
pbmc_rna_path = base::list.files(
      path = here::here("output_data"),
      full.names = TRUE
    )
pbmc_rna_path = stringr::str_subset(pbmc_rna_path,
                                    "pbmc_RNA_qc_annotated.rds")
pbmc_rna <- readRDS(pbmc_rna_path)
pbmc_rna <- UpdateSeuratObject(pbmc_rna)

# Find anchors between scATAC gene activity and scRNA gene expression
transfer.anchors <- FindTransferAnchors(
  reference = pbmc_rna,
  query = pbmc_atac,
  reduction = 'cca'
)

# Transfer cell-type labels from scRNA-seq to scATAC-seq basing on anchors
predicted.labels <- TransferData(
  anchorset = transfer.anchors,
  refdata = pbmc_rna$predicted.id,
  weight.reduction = pbmc_atac[['lsi']],
  dims = 2:30 # Exclude the first LSI component
)
pbmc_atac <- AddMetaData(object = pbmc_atac, metadata = predicted.labels)

# Since we are using a scRNA+ATAC multiome dataset, we can determine the real 
# correspondence between scRNA and scATAC cells based on their barcodes. 
# Therefore, we project the cell types from scRNA-seq to scATAC-seq according 
# to real correspondence to assess the performance of the transfer algorithm.
pbmc_atac@meta.data[,"real.labels"] <- pbmc_atac@meta.data[colnames(pbmc_atac),
                                                           "predicted.id"]
```


### [4.3.] Visualize cell type annotation
```{r}
plot1 <- DimPlot(
  object = pbmc_rna,
  group.by = 'predicted.id',
  label = TRUE,
  repel = TRUE) + 
  NoLegend() + 
  ggtitle('scRNA-seq')
plot1

plot2 <- DimPlot(
  object = pbmc_atac,
  group.by = 'predicted.id',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scATAC-seq, predicted labels')
plot2

plot3 <- DimPlot(
  object = pbmc_atac,
  group.by = 'real.labels',
  label = TRUE,
  repel = TRUE) + NoLegend() + ggtitle('scATAC-seq, real labels')
plot3
```

### [4.4.] Peak calling for each cell type
#### Cell Ranger aggregates all cell types together for peak calling, 
#### potentially overlooking cell-type specific peaks in rare cell type, 
#### such as pDCs. This oversight might affect the performance of differential
#### accessibility analysis.  We recommend  calling peaks for each cell type 
#### (see https://stuartlab.org/signac/articles/peak_calling for more 
#### information). However, this process requires installing python package 
#### macs2 and may take hours to compute. Therefore, we will not perform this 
#### step in today's course.



## [5.] Differential accessibility analysis
### [5.1.] Identify differentially accessible peaks between cell types
```{r}
Idents(pbmc_atac) <- pbmc_atac$predicted.id
DefaultAssay(pbmc_atac) <- 'peaks'
da_peaks <- FindMarkers(
  object = pbmc_atac,
  ident.1 = "CD4 Naive",
  ident.2 = "CD14+ Monocytes",
  test.use = 'wilcox', # Wilcoxon Rank Sum test
  min.pct = 0.1 # peaks detected in 10% cells in either of the two populations
)
head(da_peaks)
```

### [5.2.] Visualize differential peak
```{r}
# Show peak accessibility using violin plot
plot1 <- VlnPlot(
  object = pbmc_atac,
  features = rownames(da_peaks)[1], # Specify peak name
  pt.size = 0.1,
  idents = c("CD4 Naive", "CD14+ Monocytes")
)
plot1

# Show peak accessibility on UMAP
plot2 <- FeaturePlot(
  object = pbmc_atac,
  features = rownames(da_peaks)[1],  # Specify peak name
  pt.size = 0.1
)
plot2
```

### [5.3.] Annotate differential peaks to closest genes
```{r}
# Selecte differential peaks, log2(fold change) > 3
open_cd4naive <- rownames(da_peaks[da_peaks$avg_log2FC > 3, ])
open_cd14mono <- rownames(da_peaks[da_peaks$avg_log2FC < -3, ])

# Find closest genes
closest_genes_cd4naive <- ClosestFeature(pbmc_atac, regions = open_cd4naive)
closest_genes_cd14mono <- ClosestFeature(pbmc_atac, regions = open_cd14mono)

# Show differential peaks and closest genes
head(closest_genes_cd4naive)
head(closest_genes_cd14mono)
```

### [5.4.] Visualize genomic region
```{r}
# Select highlight regions
regions_highlight <- subsetByOverlaps(StringToGRanges(open_cd14mono), 
                                      LookupGeneCoords(pbmc_atac, "LYZ"))

# Plot coverage
CoveragePlot(
  object = pbmc_atac,
  region = "LYZ",
  region.highlight = regions_highlight,
  extend.upstream = 1000,
  extend.downstream = 5000
)
```


## [6.] Motif analysis
### [6.1.] Load Motif information from JASPAR database
```{r warning=FALSE}
# Extract TF motif from JASPAR database
pfm <- getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", 
              tax_group = 'vertebrates', 
              all_versions = FALSE)
)

# Add motif information to the seurat object
pbmc_atac <- AddMotifs(
  object = pbmc_atac,
  genome = BSgenome.Hsapiens.UCSC.hg38,
  pfm = pfm
)
```

### [6.2.] Identify motifs enriched in differential peaks
```{r}
# find enriched motifs in differential Peaks
DefaultAssay(pbmc_atac) <- 'peaks'
enriched.motifs <- FindMotifs(
  object = pbmc_atac,
  features = open_cd4naive # differential peaks identified in [5.3.]
)

# Plot enriched Motifs
MotifPlot(
  object = pbmc_atac,
  motifs = head(rownames(enriched.motifs))
)
```

### [6.3.] Computing motif activities with ChromVAR
```{r warning=FALSE}
# Run ChromVAR
pbmc_atac <- RunChromVAR(object = pbmc_atac, 
                         genome = BSgenome.Hsapiens.UCSC.hg38)
DefaultAssay(pbmc_atac) <- 'chromvar'

# Visualize motif activity of Mef2c on UMAP
p1 <- DimPlot(pbmc_atac, label = TRUE, pt.size = 0.1) + NoLegend()
p2 <- FeaturePlot(
  object = pbmc_atac,
  features = "MA0497.1", # Mef2c
  min.cutoff = 'q10',
  max.cutoff = 'q90',
  pt.size = 0.1
)
p1 + p2
```

### [6.4.] Identify differentially activity of motifs between cell types
```{r}
differential.activity <- FindMarkers(
  object = pbmc_atac,
  ident.1 = "CD4 Naive",
  ident.2 = "CD14+ Monocytes",
  only.pos = TRUE,
  mean.fxn = rowMeans,
  fc.name = "avg_diff"
)

# Visualize motifs with differentially activity
MotifPlot(
  object = pbmc_atac,
  motifs = head(rownames(differential.activity)),
  assay = 'peaks'
)
```

## [7.] Save seurat object
```{r}
base::saveRDS(
    object = pbmc_atac,
    file = paste0(here::here("output_data"),
                  "/pbmc_ATAC_advanced.rds")
  )
```


## [8.] Additional vignettes for further reading & analysis
### [8.1.] Trajectory analysis of scATAC data
#### https://stuartlab.org/signac/articles/monocle
### [8.2.] Transcription factor footprinting
#### https://stuartlab.org/signac/articles/footprint
### [8.3.] Co-accessible networks
#### https://stuartlab.org/signac/articles/cicero



```