---
title: "20241014_workflow_scRNAseq"
output: html_document
date: "2024-10-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## [0.] Prerequisites -------------------------

## [-1.0.] Avoid gcc compiler problems
```{r}
base::Sys.setenv(
  PATH = base::paste(
      "/opt/software/gcc/11.2.0/bin",
      base::Sys.getenv("PATH"),
      sep = ":"
    )
  )
base::Sys.setenv(
    CXX = "/opt/software/gcc/11.2.0/bin/g++"
  )
```

## [-1.1.] Install necessary packages
#### Missing packages: libtiff-4, lubridate, systemfonts
```{r eval=FALSE}
utils::install.packages("here")
utils::install.packages("ape") # used for BuildClusterTree in [3.5.]

utils::install.packages("tidygraph")
utils::install.packages("ggraph")
utils::install.packages("clustree") # used in [3.6.]

# utils::install.packages("tidyverse")
```

## [-1.2.] Check version of packages
```{r eval=FALSE}
utils::packageVersion("Seurat")
```

## [0.0.] Load necessary packages
```{r message=FALSE}
base::library("Seurat")
base::library("tidyverse")
```

## [1.] Read in & preprocess data -------------------------

## [1.0.] Check that directory exists
```{r}
dir <- base::paste0(
    here::here("input_data/"),
    "raw_feature_bc_matrix/"
  )
base::dir.exists(paths = dir)
```

## [1.1.] Read in data as sparse matrices
#### Data read in as dgCMatrices
```{r}
pbmc_data <- Seurat::Read10X(
    data.dir = dir,
    gene.column = 2, # column specifying gene names in features.tsv.gz
    cell.colum = 1 # column specifying the cell ID in barcodes.tsv.gz
  )
```

## [1.2.] Inspect data & calculate sparsity
#### See https://slowkow.com/notes/sparse-matrix/ for more info on how to access data in dgCMatrices
#### Sparse data most optimally saved in dgCMatrices
```{r}
utils::str(pbmc_data) # Inspect data
pbmc_data$`Gene Expression`[1:5,1:5] ##
base::dim(pbmc_data$`Gene Expression`)
100 - (base::length(pbmc_data$`Gene Expression`@x) / base::prod(pbmc_data$`Gene Expression`@Dim))*100 # Sparsity
```

## [1.3.] Create a Seurat object from the scRNAseq data
#### Seurat objects are S3 object-oriented objects, see https://adv-r.hadley.nz/s3.html
```{r}
pbmc_rna <- Seurat::CreateSeuratObject(
    counts = pbmc_data$`Gene Expression`, ## Specify scRNAseq
    assay = "RNA",
    project = "pbmc3k",
    min.cells = 3, # minimal amount of cells expressing a feature
    min.features = 200 # minimal amount of features detected per cell
  )
```

#### Alternatively create a Seurat object from both matrices
```{r eval=FALSE}
# pbmc_multi <- Seurat::CreateSeuratObject(
#     counts = pbmc_data,
#     assay = "Multi",
#     project = "pbmc3k",
#     min.cells = 3, # minimal amount of cells expressing a feature
#     min.features = 200 # minimal amount of features detected per cell
#   )
# utils::str(pbmc_multi)
```

## [1.4.] Inspect the Seurat object
#### Filtering in [1.3.] reduced the number of rows and columns
#### dimensions: 26315 10987; sparsity: 94.39873
```{r}
utils::str(pbmc_rna)
base::dim(pbmc_rna) # base::dim(pbmc_rna@assays$RNA@layers$counts) 
pbmc_rna@assays$RNA@layers$counts[1:5,1:5]
100 - (base::length(pbmc_rna@assays$RNA@layers$counts@x) / base::prod(pbmc_rna@assays$RNA@layers$counts@Dim))*100 # Sparsity
```

## [1.5.] Detect & count mitochondrial gene contribution
#### On terminal:  gunzip -c features.tsv.gz | cut -f2 | grep '^MT-' | head
```{r}
## base::dimnames(pbmc_rna@assays$RNA@features@.Data)[[1]]
feat_names <- base::rownames(pbmc_rna) ## base::colnames(pbmc_rna)
feat_names[base::grep("^MT-",feat_names)]
  
pbmc_rna[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_rna,
    pattern = "^MT-"
  )
```

## [1.6.] Inspect & visualize features, including mitochondrial gene contribution
```{r}
pbmc_rna@meta.data # Inspect new variable
base::colnames(pbmc_rna[[]])

Seurat::VlnPlot(
    object = pbmc_rna,
    features = base::colnames(pbmc_rna[[]])[2:4],
    ncol = 3
  )
base::mean(pbmc_rna@meta.data$percent.mt)

Seurat::FeatureScatter(
      object = pbmc_rna,
      feature1 = "nCount_RNA",
      feature2 = "percent.mt"
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = base::c(5,10),
      linetype = "dashed"
    )

Seurat::FeatureScatter(
      object = pbmc_rna,
      feature1 = "nCount_RNA",
      feature2 = "nFeature_RNA"
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = base::c(200L,2500L),
      linetype = "dashed"
    )
```

## [1.7.] Filter cells
#### Based on inspection in [1.6.]
#### Removes empty droplets, droplets with â‰¥ 2 cells, apoptotic cells or cells with insufficiently captured transcripts
#### dimensions: 26315 7573; sparsity: 94.87731
```{r}
pbmc_rna <- base::subset(
    x = pbmc_rna,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10
  )
base::dim(pbmc_rna)
100 - (base::length(pbmc_rna@assays$RNA@layers$counts@x) / base::prod(pbmc_rna@assays$RNA@layers$counts@Dim))*100 # Sparsity
```

# [2.] Process data - Normalize, scale, reduce dimensions -------------------------

## [2.0.] Normalization
```{r}
pbmc_rna <- Seurat::NormalizeData(
    object = pbmc_rna,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
```

## [2.1.] Inspect normalized data
```{r}
utils::str(pbmc_rna)
pbmc_rna@assays$RNA@layers$data
pbmc_rna[["RNA"]]$data
```

##<<boomark/SCTransform()>>

## [2.2.] Find most highly variably expressed genes
#### Identification of features (genes) with highest variability across all cells to apply variance-stabilizing data transformation 
```{r}
pbmc_rna <- Seurat::FindVariableFeatures(
    object = pbmc_rna,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
```

## [2.3.] Inspect variable genes
```{r}
utils::str(pbmc_rna)
pbmc_rna@assays$RNA@meta.data
  
lbls <- Seurat::VariableFeatures(object = pbmc_rna)[1:10]
raw <- Seurat::VariableFeaturePlot(
    object = pbmc_rna
  )
Seurat::LabelPoints(
      plot = raw,
      points = lbls,
      repel = TRUE
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = 1.25,
      linetype = "dashed"
    )
```

## [2.4.] Scale across dataset
#### Linear transformation for each feature (gene) across the variable genes to prepare principal component analysis (PCA)
#### Scale across all genes as default is to perform scaling only on the variable features
```{r}
pbmc_rna <- Seurat::ScaleData(
    object = pbmc_rna,
    features = base::rownames(pbmc_rna),
    verbose = FALSE
  )
```

#### Alternative is to use scaling for removal of potential sources of variation, e.g., mitochondrial reads or cell cycle phase
```{r}
# pbmc_rna <- Seurat::ScaleData(
#     object = pbmc_rna,
#     vars.to.regress = "percent.mt"
#   )
```

## [2.5.] Run principal component analysis
#### RunPCA when called on a Seurat objects operates on the scale.data slot of the assay by default
#### Seurat:::PrepDR5() function prepares internally the data to be used for PCA
```{r}
Seurat::DefaultAssay(object = pbmc_rna)
SeuratObject::Layers(object = pbmc_rna)
SeuratObject::LayerData(object = pbmc_rna, layer = "scale.data")[1:10,1:10]

pbmc_rna <- Seurat::RunPCA(
    object = pbmc_rna,
    features = Seurat::VariableFeatures(object = pbmc_rna),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
```

## [2.6.] Inspect PCA results
```{r}
utils::str(pbmc_rna)
pbmc_rna[["pca"]]
pbmc_rna@reductions$pca@feature.loadings

base::print(
    pbmc_rna[["pca"]],
    dims = 1:3,
    nfeatures = 10
  )
Seurat::VizDimLoadings(
    object = pbmc_rna,
    dims = 1:3,
    reduction = "pca",
    ncol = 3,
    nfeatures = 10
  )
Seurat::DimPlot(
      object = pbmc_rna,
      reduction = "pca",
      dims = 1:2
    ) +
  theme_bw() +
  theme(
      legend.position = "none",
      aspect.ratio = 1
    )
```

## [2.7.] Inspect dimensionality of the data
```{r}
Seurat::DimHeatmap(
    object = pbmc_rna,
    dims = 1:15,
    cells = 500, # top 500 cells
    balanced = TRUE,
    nfeatures = 10,
    slot = "scale.data"
  )

Seurat::ElbowPlot(
      object = pbmc_rna,
      ndims = 50,
      reduction = "pca"
    ) +
  geom_hline(
      yintercept = 1.3,
      linetype = "dashed"
    ) +
  geom_vline(
      xintercept = base::c(5,10),
      linetype = "dashed"
    )
```

#### Resampling method using JackStraw
#### Time consuming approach
```{r}
# pbmc_rna_ <- Seurat::JackStraw(
#     object = pbmc_rna,
#     dims = 50,
#     num.replicate = 100
#   )
# pbmc_rna_ <- Seurat::ScoreJackStraw(
#     object = pbmc_rna_,
#     dims = 1:50,
#     score.thresh = 1e-05
#   )
# Seurat::JackStrawPlot(
#     object = pbmc_rna_,
#     dims = 1:50
#   )
```

# [3.] Cell clustering -------------------------
#### Seurat performs clustering based on PCA scores of features (genes) in cells
#### Choice of included PCAs based on resampling test and Elbow plots

## [3.0.] Calculate nearest neighbors and corresponding graph
#### k-NN graph calculation based on cell distances in PCA latent space
#### refinement of graph edges based on similarity between neighborhoods
#### Uses RcppAnnoy::AnnoyEuclidean, RANN::nn2 (NNHelper) and for calculating
#### the shared nearest neighbour graph SNN_SmallestNonzero_Dist imnplemented via Rcpp (ComputeSNN)
```{r}
Seurat::DefaultAssay(object = pbmc_rna) # assay: "RNA"
SeuratObject::Embeddings(object = pbmc_rna[["pca"]]) # data.use: cell x PCA

pbmc_rna <- Seurat::FindNeighbors(
    object = pbmc_rna,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
```

## [3.1.] Inspect graph
```{r}
utils::str(pbmc_rna)
pbmc_rna@graphs
```

## [3.2.] Find clusters
#### Louvain or SLM algorithms to iteratively form groups of cells
#### with granularity of result adjusting the amount of identified clusters
```{r}
pbmc_rna <- Seurat::FindClusters(
    object = pbmc_rna,
    resolution = 0.5, # <----parameter
    verbose = FALSE
  )
```

## [3.3.] Perform non-linear dimension reduction
```{r warning=FALSE}
pbmc_rna <- Seurat::RunUMAP(
    object = pbmc_rna,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE
  )
```

## [3.4.] Visualize UMAP dimension reduction
#### UMAP preserves the local distances in favor of global relationships
```{r}
Seurat::DimPlot(
      object = pbmc_rna,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
    aspect.ratio = 1
  )
```

## [3.5.] Calculate cluster dendogram
```{r}
pbmc_rna_ <- pbmc_rna
pbmc_rna_ <- Seurat::BuildClusterTree(
    object = pbmc_rna,
    assay = "RNA",
    features = base::rownames(pbmc_rna),
    dims = 1:10,
    reduction = "pca",
    slot = "scale.data",
    verbose = FALSE
  )
utils::str(pbmc_rna_)

# replaces Seurat::PlotClusterTree
data.tree <- SeuratObject::Tool(
    object = pbmc_rna_,
    slot = "Seurat::BuildClusterTree"
  )
ape::plot.phylo(x = data.tree, direction = "downwards")
ape::nodelabels()
```

## [3.6.]
```{r}
pbmc_rna_ <- pbmc_rna
for(r in base::seq(0.1,1.2,0.1)){
  pbmc_rna_ <- Seurat::FindClusters(
      object = pbmc_rna_,
      resolution = r,
      verbose = FALSE
    )
}
pbmc_rna_@meta.data
base::debugonce(clustree::clustree)
clustree::clustree(
    x = pbmc_rna_,
    prefix = "RNA_snn_res.",
   return = "layout"
  )
```

```{r}
base::colnames(pbmc_rna_[[]])
pbmc_rna@meta.data %>%
  dplyr::

x, prefix, suffix = NULL, metadata = NULL, count_filter = 0, 
  prop_filter = 0.1, layout = c("tree", "sugiyama"), use_core_edges = TRUE, 
  highlight_core = FALSE, node_colour = prefix, node_colour_aggr = NULL, 
  node_size = "size", node_size_aggr = NULL, node_size_range = c(4, 
    15), node_alpha = 1, node_alpha_aggr = NULL, node_text_size = 3, 
  scale_node_text = FALSE, node_text_colour = "black", node_text_angle = 0, 
  node_label = NULL, node_label_aggr = NULL, node_label_size = 3, 
  node_label_nudge = -0.2, edge_width = 1.5, edge_arrow = TRUE, 
  edge_arrow_ends = c("last", "first", "both"), show_axis = FALSE, 
  return = c("plot", "graph", "layout"), ...)
```


```{r}
checkmate::assert_matrix(x, any.missing = FALSE, col.names = "unique", 
    min.cols = 2)
  checkmate::assert_character(prefix, any.missing = FALSE, 
    len = 1)
  checkmate::assert_character(suffix, any.missing = FALSE, 
    len = 1, null.ok = TRUE)
  checkmate::assert_number(count_filter, lower = 0, upper = nrow(x))
  checkmate::assert_number(prop_filter, lower = 0, upper = 1)
  checkmate::assert_data_frame(metadata, nrows = nrow(x), 
    col.names = "unique", null.ok = TRUE)
  assert_colour_node_aes("node_colour", prefix, metadata, 
    node_colour, node_colour_aggr)
  assert_numeric_node_aes("node_size", prefix, metadata, node_size, 
    node_size_aggr, 0, Inf)
  assert_numeric_node_aes("node_alpha", prefix, metadata, 
    node_alpha, node_alpha_aggr, 0, 1)
  checkmate::assert_number(node_text_size, lower = 0)
  if (!is.null(node_label)) {
    assert_node_aes("node_label", prefix, metadata, node_label, 
      node_label_aggr)
  }
  checkmate::assert_number(node_label_nudge)
  checkmate::assert_logical(scale_node_text, any.missing = FALSE, 
    len = 1)
  checkmate::assert_number(edge_width, lower = 0)
  checkmate::assert_logical(edge_arrow, any.missing = FALSE, 
    len = 1)
  layout <- match.arg(layout)
  checkmate::assert_flag(use_core_edges)
  return <- match.arg(return)
  edge_arrow_ends <- match.arg(edge_arrow_ends)
  checkmate::assert_flag(highlight_core)
  if (!is.null(suffix)) {
    colnames(x) <- gsub(suffix, "", colnames(x))
  }
  res_clean <- gsub(prefix, "", colnames(x))
  is_num <- suppressWarnings(!any(is.na(as.numeric(res_clean))))
  if (!is_num) {
    stop("The X portion of your clustering column names could not be ", 
      "converted to a number. Please check that your prefix and suffix ", 
      "are correct: prefix = '", prefix, "', suffix = '", 
      suffix, "'", call. = FALSE)
  }
  res_order <- order(as.numeric(res_clean))
  x <- x[, res_order]
  res_clean <- res_clean[res_order]
  if (!(is.null(metadata))) {
    metadata_names <- make.names(colnames(metadata))
    metadata_diff <- metadata_names != colnames(metadata)
    if (any(metadata_diff)) {
      warning("The following metadata column names will be converted from:\n", 
        paste(colnames(metadata)[metadata_diff], collapse = ", "), 
        "\n", "to:\n", paste(metadata_names[metadata_diff], 
          collapse = ", "), call. = FALSE)
      colnames(metadata) <- metadata_names
    }
  }
  node_aes_list <- list(colour = list(value = node_colour, 
    aggr = node_colour_aggr), size = list(value = node_size, 
    aggr = node_size_aggr), alpha = list(value = node_alpha, 
    aggr = node_alpha_aggr))
  if (!is.null(node_label)) {
    node_aes_list$label <- list(value = node_label, aggr = node_label_aggr)
  }
  node_aes_list <- check_node_aes_list(node_aes_list)
  graph <- build_tree_graph(x, prefix, count_filter, prop_filter, 
    metadata, node_aes_list)
  graph_attr <- igraph::graph_attr(graph)
  graph <- graph %>% tidygraph::activate("edges") %>% tidygraph::mutate(width = edge_width) %>% 
    tidygraph::group_by(.data$to) %>% tidygraph::mutate(is_core = .data$in_prop == 
    max(.data$in_prop)) %>% tidygraph::ungroup()
  if (use_core_edges) {
    layout <- graph %>% tidygraph::activate("edges") %>% 
      tidygraph::filter(.data$is_core) %>% ggraph::create_layout(layout)
    attributes(layout)$graph <- graph
  }
  else {
    layout <- ggraph::create_layout(graph, layout)
  }
  gg <- ggraph(layout)
  if (edge_arrow) {
    if (is.numeric(node_size)) {
      circle_size_end <- ifelse(edge_arrow_ends == "first", 
        0.1, node_size * 1.5)
      circle_size_start <- ifelse(edge_arrow_ends == "last", 
        0.1, node_size * 1.5)
    }
    else {
      circle_size_end <- ifelse(edge_arrow_ends == "first", 
        0.1, mean(node_size_range) * 1.5)
      circle_size_start <- ifelse(edge_arrow_ends == "last", 
        0.1, mean(node_size_range) * 1.5)
    }
    gg <- gg + geom_edge_link(arrow = arrow(length = unit(edge_width * 
      5, "points"), ends = edge_arrow_ends), end_cap = circle(circle_size_end, 
      "points"), start_cap = circle(circle_size_start, 
      "points"), aes(colour = .data$count, alpha = .data$in_prop, 
      edge_width = .data$is_core))
  }
  else {
    gg <- gg + geom_edge_link(aes(colour = .data$count, 
      alpha = .data$in_prop, edge_width = .data$is_core))
  }
  if (highlight_core) {
    core_width <- edge_width * 2
    gg <- gg + scale_edge_width_manual(values = c(edge_width, 
      core_width))
  }
  else {
    gg <- gg + scale_edge_width_manual(values = c(edge_width, 
      edge_width), guide = "none")
  }
  gg <- gg + scale_edge_colour_gradientn(colours = viridis::viridis(256)) + 
    scale_edge_alpha(limits = c(0, 1))
  gg <- gg + add_node_points(graph_attr$node_colour, graph_attr$node_size, 
    graph_attr$node_alpha, names(igraph::vertex_attr(graph)))
  if (scale_node_text && !is.numeric(node_size)) {
    gg <- gg + geom_node_text(aes(label = .data$cluster, 
      size = .data[[graph_attr$node_size]]), colour = node_text_colour, 
      angle = node_text_angle)
  }
  else {
    gg <- gg + geom_node_text(aes(label = .data$cluster), 
      size = node_text_size, colour = node_text_colour, 
      angle = node_text_angle)
  }
  if (!(is.null(node_label))) {
    gg <- gg + add_node_labels(graph_attr$node_label, graph_attr$node_colour, 
      node_label_size, node_text_colour, node_label_nudge, 
      names(igraph::vertex_attr(graph)))
  }
  gg <- gg + scale_size(range = node_size_range) + ggraph::theme_graph(base_family = "", 
    plot_margin = ggplot2::margin(2, 2, 2, 2))
  if (show_axis) {
    gg <- gg + ylab(prefix) + scale_y_continuous(breaks = sort(unique(layout$y)), 
      labels = rev(res_clean)) + theme(axis.text.y = element_text(), 
      axis.title = element_text(), axis.title.x = element_blank(), 
      panel.grid.major.y = element_line(colour = "grey92"))
  }
  if (return == "plot") {
    return(gg)
  }
  else if (return == "graph") {
    return(graph)
  }
  else if (return == "layout") {
    return(layout)
```

