---
title: "Novo Nordisk Applied Single Cell Analysis course October 2024 - scRNAseq analysis advanced"
output: html_document
date: "2024-10-17"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---

## [0.] Prerequisites

### [0.0.] Read in data
#### See [1.0.] - [1.1.] in '20241014_scRNAseq_analysis.Rmd'.
```{r}
dir <- base::paste0(
    here::here("input_data/"),
    "raw_feature_bc_matrix/"
  )
pbmc_data <- Seurat::Read10X(
    data.dir = dir,
    gene.column = 2,
    cell.colum = 1
  )
```

#### See [6.0.] in '20241014_scRNAseq_analysis.Rmd'.
```{r}
pbmc_rna <- base::list.files(
      path = here::here("output_data"),
      full.names = TRUE
    ) %>%
  stringr::str_subset("RNA.rds") %>%
  base::readRDS(
      file = .
    )
```

### [0.1.] Load necessary packages
```{r message=FALSE}
base::library("Seurat")
base::library("tidyverse")
```

---

## [4.] Additional analysis & quality checks

### [4.0.] Subsample number of cells in initial scRNAseq features x cells matrix
#### 100,000 instead of 733,612 input cells.
```{r}
base::set.seed(1)
sub = 1e+5
cols <- base::sample(
    x = 1:base::ncol(pbmc_data$`Gene Expression`),
    size = sub,
    replace = FALSE
  )
pbmc_alt_data <- pbmc_data$`Gene Expression`[,cols]
```

### [4.1.] Run analysis with subsampled cells
#### No parameters changed comapred to [1.3.] - [3.4.] in '20241014_scRNAseq_analysis.Rmd'.
```{r}
base::set.seed(1)
pbmc_alt <- Seurat::CreateSeuratObject(
    counts = pbmc_alt_data,
    assay = "RNA",
    project = "pbmc10k",
    min.cells = 3,
    min.features = 200
  )
pbmc_alt[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_alt,
    pattern = "^MT-"
  )
pbmc_alt <- base::subset(
    x = pbmc_alt,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10
  )
pbmc_alt <- Seurat::NormalizeData(
    object = pbmc_alt,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
pbmc_alt <- Seurat::FindVariableFeatures(
    object = pbmc_alt,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
pbmc_alt <- Seurat::ScaleData(
    object = pbmc_alt,
    features = base::rownames(pbmc_alt),
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunPCA(
    object = pbmc_alt,
    features = Seurat::VariableFeatures(object = pbmc_alt),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
pbmc_alt <- Seurat::FindNeighbors(
    object = pbmc_alt,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
pbmc_alt <- Seurat::FindClusters(
    object = pbmc_alt,
    resolution = 0.5,
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunUMAP(
    object = pbmc_alt,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE
  )
Seurat::DimPlot(
      object = pbmc_alt,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
base::remove(pbmc_alt,pbmc_alt_data)
```

---

### [4.2.] Subsample reads in gene expression matrix
#### From 38.067.789 to 10.000.000 or 5.000.000.
#### See https://github.com/barkasn/nbHelpers/blob/master/R/subSampleCountMatrix.R
#### for 'nbHelpers::subSampleCountMatrix2()' function.
```{r}
base::set.seed(1)
fin = 5e6 # 1e+7
mat <- pbmc_data$`Gene Expression`
cur <- base::sum(mat@x)
rmv <- cur - fin
smp <- base::sample(
    x = base::sum(mat@x),
    size = rmv,
    replace = FALSE
  )
pos <- base::cumsum(mat@x)
t <- base::table(
    base::findInterval(
      x = smp,
      vec = pos,
      rightmost.closed = FALSE,
      left.open = TRUE
    ) + 1
  )
r <- base::as.numeric(t)
base::names(r) <- base::names(t)
t <- r
n <- t[base::match(base::as.character(1:base::length(mat@x)), base::names(t))]
n[base::is.na(n)] <- 0
mat@x <- mat@x - n # remove sampled reads
pbmc_alt_data <- Matrix::drop0(mat) # drop new 0s
base::sum(pbmc_alt_data@x)
base::remove(r,t,mat,cur,rmv,smp,pos,n)
```

### [4.3.] Run analysis with subsampled reads
#### No parameters changed comapred to [1.3.] - [3.4.] in '20241014_scRNAseq_analysis.Rmd'.
```{r}
base::set.seed(1)
pbmc_alt <- Seurat::CreateSeuratObject(
    counts = pbmc_alt_data,
    assay = "RNA",
    project = "reads10mio",
    min.cells = 3,
    min.features = 200
  )
pbmc_alt[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_alt,
    pattern = "^MT-"
  )
pbmc_alt <- base::subset(
    x = pbmc_alt,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10
  )
pbmc_alt <- Seurat::NormalizeData(
    object = pbmc_alt,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
pbmc_alt <- Seurat::FindVariableFeatures(
    object = pbmc_alt,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
pbmc_alt <- Seurat::ScaleData(
    object = pbmc_alt,
    features = base::rownames(pbmc_alt),
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunPCA(
    object = pbmc_alt,
    features = Seurat::VariableFeatures(object = pbmc_alt),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
pbmc_alt <- Seurat::FindNeighbors(
    object = pbmc_alt,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
pbmc_alt <- Seurat::FindClusters(
    object = pbmc_alt,
    resolution = 0.5,
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunUMAP(
    object = pbmc_alt,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE
  )
Seurat::DimPlot(
      object = pbmc_alt,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
    aspect.ratio = 1
  )
base::remove(pbmc_alt,pbmc_alt_data)
```

---

### [4.4.] Read out metainfo & UMAP data
```{r}
assertthat::are_equal(
    x = base::rownames(pbmc_rna@meta.data),
    y = base::rownames(pbmc_rna@reductions$umap@cell.embeddings)
  )

meta <- pbmc_rna@meta.data %>%
  tibble::as_tibble(rownames = "cell")
umap <- pbmc_rna@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell")

meta_comb <- dplyr::left_join(
    x = umap,
    y = meta,
    by = "cell"
  )
```

### [4.5.] Average mean cluster positions for cluster labels
```{r}
cls <- meta_comb %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::summarise(
      pos_1 = base::mean(umap_1),
      pos_2 = base::mean(umap_2),
      .groups = "drop"
    )
```

### [4.6.] Map quality metrics onto UMAP
```{r}
coef = 0.5
cols <- base::c(
    grDevices::colorRampPalette(
        base::c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e")
      )(coef*100),
    grDevices::colorRampPalette(
      base::c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"),
      bias=2
    )((1-coef)*100)
  )

ggplot2::ggplot(
      data = meta_comb,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        color = nFeature_RNA # percent.mt # nCount_RNA, 
      )
    ) +
    geom_point(
      size = 0.1
    ) +
  ggrepel::geom_label_repel(
      inherit.aes = FALSE,
      data = cls,
      mapping = aes(
          x = pos_1,
          y = pos_2,
          label = seurat_clusters
        )
    ) +
  scale_colour_gradientn(
      colours = cols
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      axis.text = element_blank(),
      axis.ticks = element_blank()
    ) +
  guides(
    fill = "none",
    colour = guide_colourbar(
        label.position = "right",
        title = NULL,
        title.position = "left",
        title.theme = element_text(
            angle = 90,
            hjust = 0.5
          ),
        ticks = T,
        nbin = 100,
        barheight = 8,
        label = T,
        barwidth = 0.5
      )
    )
```

### [4.7.] Cluster-specific quality metrics
```{r}
ggplot2::ggplot(
      data = meta_comb,
      mapping = aes(
        x = seurat_clusters,
        y = nFeature_RNA # percent.mt # nCount_RNA
      )
    ) +
  geom_jitter(
      alpha = 0.1,
      height = 0,
      width = 0.3
    ) +
  geom_boxplot(
      outliers = FALSE,
      alpha = 0.5
    ) +
  geom_hline(
      yintercept = 750L,
      linetype = "dashed"
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 2
    )
```

---

### [4.8.] Rerun analysis by changing QC cut-offs
#### No parameters changed comapred to [1.3.] - [3.4.] in '20241014_scRNAseq_analysis.Rmd'.
#### Elbow plot looks identical to pbmc_rna.
```{r}
pbmc_qc <- Seurat::CreateSeuratObject(
    counts = pbmc_data$`Gene Expression`,
    assay = "RNA",
    project = "pbmc10k",
    min.cells = 3,
    min.features = 200
  )
pbmc_qc[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_qc,
    pattern = "^MT-"
  )
pbmc_qc <- base::subset(
    x = pbmc_qc,
    subset = nFeature_RNA > 750 & nFeature_RNA < 2500 & percent.mt < 10
  )
pbmc_qc <- Seurat::NormalizeData(
    object = pbmc_qc,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
pbmc_qc <- Seurat::FindVariableFeatures(
    object = pbmc_qc,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
pbmc_qc <- Seurat::ScaleData(
    object = pbmc_qc,
    features = base::rownames(pbmc_qc),
    verbose = FALSE
  )
pbmc_qc <- Seurat::RunPCA(
    object = pbmc_qc,
    features = Seurat::VariableFeatures(object = pbmc_qc),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
pbmc_qc <- Seurat::FindNeighbors(
    object = pbmc_qc,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
pbmc_qc <- Seurat::FindClusters(
    object = pbmc_qc,
    resolution = 0.3,
    verbose = FALSE,
    random.seed = 0L
  )
pbmc_qc <- Seurat::RunUMAP(
    object = pbmc_qc,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE,
    seed.use = 42L
  )
Seurat::DimPlot(
      object = pbmc_qc,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

### [4.9.] Map e.g., mitochondrial contribution to UMAP
```{r}
meta_qc <- pbmc_qc@meta.data %>%
  tibble::as_tibble(rownames = "cell")
umap_qc <- pbmc_qc@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell")
meta_qc_comb <- dplyr::left_join(
    x = umap_qc,
    y = meta_qc,
    by = "cell"
  )

cls <- meta_qc_comb %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::summarise(
      pos_1 = base::mean(umap_1),
      pos_2 = base::mean(umap_2),
      .groups = "drop"
    )

ggplot2::ggplot(
      data = meta_qc_comb,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        color = nFeature_RNA # percent.mt # nCount_RNA
      )
    ) +
  geom_point(
      size = 0.1
    ) +
  ggrepel::geom_label_repel(
      inherit.aes = FALSE,
      data = cls,
      mapping = aes(
          x = pos_1,
          y = pos_2,
          label = seurat_clusters
        )
    ) +
  scale_colour_gradientn(
      colours = cols
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      axis.text = element_blank(),
      axis.ticks = element_blank()
    ) +
  guides(
    fill = "none",
    colour = guide_colourbar(
        label.position = "right",
        title = NULL,
        title.position = "left",
        title.theme = element_text(
            angle = 90,
            hjust = 0.5
          ),
        ticks = T,
        nbin = 100,
        barheight = 8,
        label = T,
        barwidth = 0.5
      )
    )
```

### [4.10.] Mean e.g., mitochondrial contribution per cluster
```{r}
ggplot2::ggplot(
      data = meta_qc_comb,
      mapping = aes(
        x = seurat_clusters,
        y = nCount_RNA # percent.mt # nFeature_RNA
      )
    ) +
  geom_jitter(
      alpha = 0.1,
      height = 0,
      width = 0.3
    ) +
  geom_boxplot(
      outliers = FALSE,
      alpha = 0.5
    ) +
  theme_bw() +
  theme(
    aspect.ratio = 2
  )
```

---

#### Question: Do cluster 7 & 8 consist of doublets or are these biologically meaningful cell types?
#### (e.g., annotations, doublet detection)

---

## [5.] Marker genes

### [5.0.] Identify all differential expressions
#### Check with 'base::range(pbmc_markers$avg_log2FC)' and 'base::range(pbmc_markers$p_val)'.
```{r}
pbmc_markers <- Seurat::FindAllMarkers(
    object = pbmc_qc,
    only.pos = TRUE,
    assay = "RNA",
    slot = "data", # "scale.data"
    test.use = 'wilcox',
    verbose = FALSE,
    random.seed = 1L,
    return.thresh = 1, # no limitation
    logfc.threshold = -Inf # no limitation
  )
```

### [5.1.] Inspect markers
#### Matching of canonical cell type markers to identified unbiased marker genes.
```{r}
pbmc_markers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(
      p_val_adj < 1e-2 & avg_log2FC > 1
    ) %>%
  dplyr::summarise(
      nFeat_sigDE = dplyr::n()
    )

pbmc_markers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(
      p_val_adj < 1e-2
    ) %>%
  dplyr::arrange(cluster, dplyr::desc(avg_log2FC)) %>%
  dplyr::filter(
      dplyr::row_number() %in% 1:10
    )
```

### [5.2.] Read out gene expression values from Seurat object
#### Default plotting via 'Seurat::FeaturePlot(pbmc_qc, features = "CD8A")'.
#### Access to genes via 'SeuratObject::Features(x = pbmc_qc,layer = "data")'.
```{r eval=FALSE}
gene <- "CD8A"

Seurat::FetchData(
    object = pbmc_qc,
    vars = base::c(gene,"umap_1","umap_2"),
    cells = base::colnames(x = pbmc_qc),
    assay = "RNA",
    layer = "data"
  )

tibble::enframe(
  SeuratObject::LayerData(
      object = pbmc_qc,
      layer = "data",
      cells = base::colnames(x = pbmc_qc),
      features = base::rownames(x = pbmc_qc)
    )[gene,]
  )
```

#### Use direct access to the S3 Seurat object data.
```{r}
gene <- "CD8A"
m <- pbmc_qc@assays$RNA@layers$data
base::colnames(m) <- base::colnames(pbmc_qc)
base::rownames(m) <- base::rownames(pbmc_qc)
ge <- m[gene,]
ge_qc <- tibble::enframe(
    x = ge, name = "cell", value = gene
  )
umap_qc <- pbmc_qc@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell")
ge_comb <- dplyr::left_join(
    x = umap_qc, y = ge_qc, by = "cell"
  )
```

### [5.3.] Plot gene expression values onto the UMAP projection
```{r}
coef = 0.5
cols <- base::c(
    grDevices::colorRampPalette(
        base::c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e")
      )(coef*100),
    grDevices::colorRampPalette(
      base::c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"),
      bias=2
    )((1-coef)*100)
  )

ggplot2::ggplot(
      data = ge_comb,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        color = CD8A
      )
    ) +
  geom_point(
      size = 0.1
    ) +
  scale_colour_gradientn(
      colours = cols
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      axis.text = element_blank(),
      axis.ticks = element_blank()
    )
```

### [5.4.] Read in reference data
```{r message=FALSE, warning=FALSE}
ref <- base::readRDS(
    file = base::paste0(
      here::here("input_data/"),
      "pbmc_10k_v3.rds"
    )
  )
ref <- Seurat::UpdateSeuratObject(object = ref)
```

### [5.5.] Inspect reference dataset
```{r}
utils::str(ref)
ref@meta.data
Seurat::DimPlot(
      object = ref,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

### [5.6.] Transfer annotations via identified anchors
```{r}
pbmc_anchors <- Seurat::FindTransferAnchors(
    reference = ref,
    query = pbmc_qc,
    dims = 1:50,
    reference.reduction = "pca",
    verbose = FALSE
  )
predictions <- Seurat::TransferData(
    anchorset = pbmc_anchors,
    refdata = ref$celltype,
    dims = 1:50,
    verbose = FALSE
  )
pbmc_qc <- Seurat::AddMetaData(
    object = pbmc_qc,
    metadata = predictions
  )
```

### [5.7.] Inspect & plot data
```{r}
pbmc_qc@meta.data
Seurat::DimPlot(
      object = pbmc_qc,
      reduction = "umap",
      group.by = "predicted.id"
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

### [5.8.] Label predicted IDs onto the UMAP dimension reduction
```{r}
pal <- base::c(
    "#3255A4", "#A4DBE8", "#8C8279",
    "#EA7600", "#F6BE00", "#28B463",
    "#AF7AC5", "#0E6655", "#0097A9",
    "#E03C31", "#B5BD00", "#500778",
    "#93272C", "#C6B0BC", "#8F993E",
    "#17202A", "#FF6F00", "#555025"
  )

meta_qc <- pbmc_qc@meta.data %>%
  tibble::as_tibble(rownames = "cell")
umap_qc <- pbmc_qc@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell")
meta_qc_comb <- dplyr::left_join(
    x = umap_qc,
    y = meta_qc,
    by = "cell"
  )

cls <- meta_qc_comb %>%
  dplyr::group_by(predicted.id) %>%
  dplyr::summarise(
      pos_1 = base::mean(umap_1),
      pos_2 = base::mean(umap_2),
      .groups = "drop"
    )

ggplot2::ggplot(
      data = meta_qc_comb,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        color = predicted.id
      )
    ) +
  geom_point(
      size = 0.1
    ) +
  ggrepel::geom_label_repel(
      inherit.aes = FALSE,
      data = cls,
      mapping = aes(
          x = pos_1,
          y = pos_2,
          label = predicted.id
        ),
      box.padding = 1,
      max.overlaps = Inf
    ) +
  scale_colour_manual(
      values = pal
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      axis.text = element_blank(),
      axis.ticks = element_blank()
    )
```

---

#### Question: Are platelets a real cluster or ambient RNA?

```{r}
meta_qc_comb %>%
  dplyr::filter(predicted.id == "Platelet")
```

---

### [5.9.] Plot significantly enriched pathways between two cell types
```{r}
dplyr::distinct(meta_qc_comb,predicted.id,seurat_clusters)
base::library(enrichR)
enrichR::listEnrichrSites()
enrichR::setEnrichrSite("Enrichr")
dbs <- enrichR::listEnrichrDbs()

Seurat::DEenrichRPlot(
    object = pbmc_qc,
    ident.1 = base::c(5),
    ident.2 = base::c(1,7),
    balanced = TRUE,
    logfc.threshold = 0.25,
    assay = "RNA",
    layer = "data",
    test.use = "wilcox",
    p.val.cutoff = 0.05,
    cols = pal,
    enrich.database = "GO_Molecular_Function_2015",
    max.genes = base::length(base::rownames(pbmc_qc))
  )
```

---

### [5.10.] SCTransform
#### No QC filtering performed, which leads to inclusion of contaminations again.
```{r}
base::set.seed(1L)
pbmc_sct <- Seurat::CreateSeuratObject(
    counts = pbmc_data$`Gene Expression`,
    assay = "RNA",
    project = "pbmcSCT",
    min.cells = 3,
    min.features = 200
  )
pbmc_sct[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_sct,
    pattern = "^MT-"
  )
# Necessary package: BiocManager::install('glmGamPoi')
pbmc_sct <- Seurat::SCTransform(
    object = pbmc_sct,
    verbose = FALSE,
    # vars.to.regress = "percent.mt"
  )
utils::str(pbmc_sct)
pbmc_sct <- Seurat::RunPCA(
    object = pbmc_sct,
    features = Seurat::VariableFeatures(object = pbmc_sct),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "SCT",
    layer = 'scale.data' # default
  )
pbmc_sct <- Seurat::FindNeighbors(
    object = pbmc_sct,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
pbmc_sct <- Seurat::FindClusters(
    object = pbmc_sct,
    resolution = 0.3,
    verbose = FALSE
  )
pbmc_sct <- Seurat::RunUMAP(
    object = pbmc_sct,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE
  )
Seurat::DimPlot(
      object = pbmc_sct,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

---

#### Question: Does SCTransform recapitulate the cell type distribution identified via transferred annotations better?
#### (e.g., better separation of small clusters like 'Double negative T-cells')

---

## [6.] Save & load data

### [6.1.] Save the "nFeature_RNA"-cleaned and annotated version pbmc_qc
```{r eval=FASLE}
base::saveRDS(
    object = pbmc_qc,
    file = base::paste0(
        here::here("output_data/"),
        base::format(base::Sys.time(),"%Y%m%d"),
        "_pbmc_RNA_qc_annotated",
        ".rds"
      ),
    compress = TRUE
  )
```

```{r}
pbmc_qc <- base::list.files(
      path = here::here("output_data"),
      full.names = TRUE
    ) %>%
  stringr::str_subset("RNA_qc.rds") %>%
  base::readRDS(
      file = .
    )
utils::str(pbmc_qc)
```
