---
title: "20251229_scRNAseq_analysis_v3"
output: html_document
author: "Krautz, Robert"
date: "2025-12-29"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## [0.] Prerequisites

## [0.0.] Install necessary packages
```{r eval=FALSE}
utils::install.packages("here")
if(base::dir.exists(base::find.package("ggplot2"))){
    utils::remove.packages("ggplot2")
  }
utils::install.packages(
    "https://cran.r-project.org/src/contrib/Archive/ggplot2/ggplot2_3.4.4.tar.gz",
    type = "source",
    repos = NULL
  )
if(base::dir.exists(base::find.package("patchwork"))){
    utils::remove.packages("patchwork")
  }
utils::install.packages(
    "https://cran.r-project.org/src/contrib/Archive/patchwork/patchwork_1.1.3.tar.gz",
    type = "source",
    repos = NULL
  )
if(base::dir.exists(base::find.package("cowplot"))){
    utils::remove.packages("cowplot")
  }
utils::install.packages(
    "https://cran.r-project.org/src/contrib/Archive/cowplot/cowplot_1.1.1.tar.gz",
    type = "source",
    repos = NULL
  )

utils::install.packages(
    "https://cran.r-project.org/src/contrib/Archive/Seurat/Seurat_5.0.0.tar.gz",
    type = "source",
    repos = NULL
  )
utils::install.packages("tidyverse", force = TRUE)
utils::install.packages("scales", force = TRUE)
utils::install.packages("ggplot2", force = TRUE)
# utils::install.packages("Seurat", force = TRUE)
utils::install.packages("enrichR")
devtools::install_github('immunogenomics/presto')
utils::install.packages("assertthat")
```

```{r eval=FALSE}
utils::install.packages("ape") # used for BuildClusterTree in [3.5.]
utils::install.packages("tidygraph")
utils::install.packages("ggraph")
utils::install.packages("clustree") # used in [3.6.]
utils::install.packages("WriteXLS")
utils::install.packages("ellipsis")
utils::install.packages("pkgbuild")
utils::install.packages("pkgload")
utils::install.packages("profvis")
utils::install.packages("remotes")
utils::install.packages("sessioninfo")
utils::install.packages("urlchecker")
utils::install.packages("usethis")
devtools::install_github("wjawaid/enrichR")
```

### [0.1.] Check version of packages
#### Worfklow relies on Seurat v5 (v5.0.0 or v5.1.0).
```{r eval=FALSE}
utils::packageVersion("Seurat") # 5.0.0
utils::packageVersion("tidyverse") # 2.0.0
utils::packageVersion("ggplot2") # 3.4.4
utils::packageVersion("patchwork") # 1.1.3
utils::packageVersion("cowplot") # 1.2.0
```

### [0.3.] Load necessary packages
```{r message=FALSE}
base::library("Seurat")
base::library("tidyverse")
base::library("enrichR")
```


## [1.] Read in & preprocess data

### [1.0.] Check that directory exists
```{r}
dir <- base::paste0(
    here::here("input_data/"),
    "raw_feature_bc_matrix/"
  )
base::dir.exists(paths = dir)
```

### [1.1.] Read in data as sparse matrices
#### Cellranger provides 'barcodes', 'features', 'matrix' output files
#### Data read in as dgCMatrices.
```{r}
pbmc_data <- Seurat::Read10X(
    data.dir = dir,
    gene.column = 2, # column specifying gene names in features.tsv.gz
    cell.colum = 1 # column specifying the cell ID in barcodes.tsv.gz
  )
```

### [1.2.] Inspect data & calculate sparsity
#### See https://slowkow.com/notes/sparse-matrix/ for more info on how to access data in dgCMatrices.
#### Sparse data most optimally saved in dgCMatrices.
#### Dimensions: 36601 x 733612, Sparsity: 99.92142
```{r}
# utils::str(pbmc_data)
pbmc_data$`Gene Expression`[1:5,1:5]
base::dim(pbmc_data$`Gene Expression`)
100 - (base::length(pbmc_data$`Gene Expression`@x) / base::prod(pbmc_data$`Gene Expression`@Dim))*100 # Sparsity
```

### [1.3.] Create a Seurat object from the scRNAseq data
#### Seurat objects are S3 object-oriented objects, see https://adv-r.hadley.nz/s3.html.
```{r}
pbmc_rna <- Seurat::CreateSeuratObject(
    counts = pbmc_data$`Gene Expression`, ## Specify scRNAseq
    assay = "RNA",
    project = "pbmc",
    min.cells = 3, # minimal amount of cells expressing a feature
    min.features = 200 # minimal amount of features detected per cell
  )

utils::str(pbmc_rna)
pbmc_rna@assays # see Layers
```

#### Alternatively create a Seurat object from both matrices.
```{r eval=FALSE}
pbmc_multi <- Seurat::CreateSeuratObject(
    counts = pbmc_data,
    assay = "Multi",
    project = "pbmc",
    min.cells = 3, # minimal amount of cells expressing a feature
    min.features = 200 # minimal amount of features detected per cell
  )

utils::str(pbmc_multi)
pbmc_multi@assays # see Layers
```

### [1.4.] Inspect the Seurat object
#### Filtering in [1.3.] reduced the number of rows and columns.
#### Dimensions: 26315 x 10987, Sparsity: 94.39873
```{r eval=FALSE}
utils::str(pbmc_rna)
base::dim(pbmc_rna)
base::dim(pbmc_rna@assays$RNA@layers$counts)
pbmc_rna@assays$RNA@layers$counts[1:5,1:5]
100 - (base::length(pbmc_rna@assays$RNA@layers$counts@x) / base::prod(pbmc_rna@assays$RNA@layers$counts@Dim))*100 # Sparsity
```

### [1.5.] Detect & count mitochondrial gene contribution
#### On terminal:  gunzip -c features.tsv.gz | cut -f2 | grep '^MT-' | head
```{r}
feat_names <- base::rownames(pbmc_rna) 
feat_names[base::grep("^MT-",feat_names)]
  
pbmc_rna[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_rna,
    pattern = "^MT-"
  )
```

### [1.6.] Inspect & visualize features, including mitochondrial gene contribution
```{r}
base::mean(pbmc_rna@meta.data$percent.mt)

Seurat::FeatureScatter(
      object = pbmc_rna,
      feature1 = "nCount_RNA",
      feature2 = "percent.mt"
    ) +
  ggplot2::geom_density2d() +
  geom_hline(
      yintercept = base::c(5,10),
      linetype = "dashed"
    )

Seurat::FeatureScatter(
      object = pbmc_rna,
      feature1 = "nCount_RNA",
      feature2 = "nFeature_RNA"
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = base::c(200L,2500L),
      linetype = "dashed"
    )
```

```{r echo=FALSE, eval=FALSE}
pbmc_rna@meta.data %>%
  arrange(percent.mt) %>%
ggplot2::ggplot(
      data = .,
      mapping = aes(
        x = nCount_RNA,
        y = nFeature_RNA,
        colour = percent.mt
      )
    )+
  geom_point() +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

### [1.7.] Filter cells
#### Based on inspection in [1.6.].
#### Removes empty droplets, droplets with â‰¥ 2 cells, apoptotic cells or cells with insufficiently captured transcripts.
#### Dimensions: 26315 x 7573, Sparsity: 94.87731
```{r}
pbmc_rna <- base::subset(
    x = pbmc_rna,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10
  )

100 - (base::length(pbmc_rna@assays$RNA@layers$counts@x) / base::prod(pbmc_rna@assays$RNA@layers$counts@Dim))*100 # Sparsity
```

## [2.] Process data - Normalize, scale, reduce dimensions

### [2.0.] Normalization
```{r}
pbmc_rna <- Seurat::NormalizeData(
    object = pbmc_rna,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
```

### [2.1.] Inspect normalized data
```{r eval=FALSE}
utils::str(pbmc_rna)
pbmc_rna@assays$RNA@layers$data
pbmc_rna[["RNA"]]$data
```

### [2.2.] Find most highly variably expressed genes
#### Identification of features (genes) with highest variability across all cells to apply variance-stabilizing data transformation.
```{r}
pbmc_rna <- Seurat::FindVariableFeatures(
    object = pbmc_rna,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
```

### [2.3.] Inspect variable genes
```{r}
lbls <- Seurat::VariableFeatures(object = pbmc_rna)[1:10]
raw <- Seurat::VariableFeaturePlot(
    object = pbmc_rna
  )
Seurat::LabelPoints(
      plot = raw,
      points = lbls,
      repel = TRUE,
      xnudge = 0,
      ynudge = 0
    ) +
  geom_density2d() +
  geom_hline(
      yintercept = 1.25,
      linetype = "dashed"
    )
```

### [2.4.] Scale across dataset
#### Linear transformation for each feature (gene) across the variable genes to prepare principal component analysis (PCA).
#### Scale across all genes as default is to perform scaling only on the variable features.
```{r}
pbmc_rna <- Seurat::ScaleData(
    object = pbmc_rna,
    features = base::rownames(pbmc_rna),
    verbose = FALSE
  )
```

#### Alternative is to use scaling for removal of potential sources of variation, e.g., mitochondrial reads or cell cycle phase.
```{r eval=FALSE}
pbmc_rna <- Seurat::ScaleData(
    object = pbmc_rna,
    vars.to.regress = "percent.mt"
  )
```

### [2.5.] Run principal component analysis
#### RunPCA when called on a Seurat objects operates on the scale.data slot of the assay by default.
#### Seurat:::PrepDR5() function prepares internally the data to be used for PCA.
```{r}
pbmc_rna <- Seurat::RunPCA(
    object = pbmc_rna,
    features = Seurat::VariableFeatures(object = pbmc_rna),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 5L, # 1L
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
```

### [2.6.] Inspect PCA results
```{r}
base::print(
    pbmc_rna[["pca"]],
    dims = 1:3,
    nfeatures = 10
  )

pbmc_rna@reductions$pca@cell.embeddings %>%
  tibble::as_tibble() %>% 
  dplyr::select("PC_1","PC_2") %>%
  ggplot2::ggplot(
      data = .,
      mapping = aes(
        x = PC_1,
        y = PC_2
      )
    ) +
  geom_point(
      size = 0.1
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

### [2.7.] Inspect dimensionality of the data
```{r}
Seurat::DimHeatmap(
    object = pbmc_rna,
    dims = 1:15,
    cells = 500, # top 500 cells
    balanced = TRUE,
    nfeatures = 10,
    slot = "scale.data"
  )

Seurat::ElbowPlot(
      object = pbmc_rna,
      ndims = 50,
      reduction = "pca"
    ) +
  geom_hline(
      yintercept = 1.3,
      linetype = "dashed"
    ) +
  geom_vline(
      xintercept = base::c(5,10),
      linetype = "dashed"
    )
```

### [2.8.] Resampling method using JackStraw.
#### Not used or recommended as it is a time-consuming approach.
```{r eval = FALSE}
pbmc_rna_ <- Seurat::JackStraw(
    object = pbmc_rna,
    dims = 50,
    num.replicate = 100
  )
pbmc_rna_ <- Seurat::ScoreJackStraw(
    object = pbmc_rna_,
    dims = 1:50,
    score.thresh = 1e-05
  )
Seurat::JackStrawPlot(
    object = pbmc_rna_,
    dims = 1:50
  )
```

## [3.] Cell clustering
#### Seurat performs clustering based on PCA scores of features (genes) in cells.
#### Choice of included PCAs based on resampling test and Elbow plots.

### [3.0.] Calculate nearest neighbors and corresponding graph
#### k-NN graph calculation based on cell distances in PCA latent space.
#### Refinement of graph edges based on similarity between neighborhoods.
#### Uses RcppAnnoy::AnnoyEuclidean and RANN::nn2 (NNHelper) for calculating.
#### The shared nearest neighbour graph SNN_SmallestNonzero_Dist imnplemented via Rcpp (ComputeSNN).
```{r}
pbmc_rna <- Seurat::FindNeighbors(
    object = pbmc_rna,
    dims = 1:10, # 1:10 / 1:5 / 1:4
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
```

### [3.1.] Inspect graph
```{r eval=FALSE}
utils::str(pbmc_rna)
pbmc_rna@graphs
```

### [3.2.] Find clusters
#### Louvain or SLM algorithms to iteratively form groups of cells.
#### Granularity of result adjusting the amount of identified clusters.
```{r}
pbmc_rna <- Seurat::FindClusters(
    object = pbmc_rna,
    resolution = 0.1, # <---- granularity parameter, original res 0.3
    verbose = FALSE,
    random.seed = 0
  )
```

### [3.3.] Perform non-linear dimension reduction
```{r warning=FALSE}
pbmc_rna <- Seurat::RunUMAP(
    object = pbmc_rna,
    umap.method = "uwot",
    dims = 1:10, # 1:10 / 1:5 / 1:4
    verbose = FALSE,
    seed.use = 42
  )
```

### [3.4.] Visualize UMAP dimension reduction
#### UMAP preserves the local distances in favor of global relationships.
```{r}
cls <- pbmc_rna@meta.data %>%
  tibble::as_tibble(rownames = "cell") %>%
  dplyr::select(cell,seurat_clusters)

cls_umap <- pbmc_rna@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell") %>% 
  dplyr::left_join(cls, by = "cell")

cls_mid <- cls_umap %>% 
  dplyr::group_by(seurat_clusters) %>%
  dplyr::summarise(
      pos_1 = base::mean(umap_1),
      pos_2 = base::mean(umap_2),
      .groups = "drop"
    )

ggplot2::ggplot(
      data = cls_umap,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        colour = seurat_clusters
      )
    ) +
  geom_point(
      size = 0.1
    ) +
   ggrepel::geom_label_repel(
      inherit.aes = FALSE,
      data = cls_mid,
      mapping = aes(
          x = pos_1,
          y = pos_2,
          label = seurat_clusters
        )
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      legend.position = "none"
    )
```

### [3.5.] Calculate cluster dendogram
```{r}
pbmc_rna_ <- pbmc_rna
pbmc_rna_ <- Seurat::BuildClusterTree(
    object = pbmc_rna,
    assay = "RNA",
    features = base::rownames(pbmc_rna),
    dims = 1:10,
    reduction = "pca",
    slot = "scale.data",
    verbose = FALSE
  )

data.tree <- SeuratObject::Tool(
    object = pbmc_rna_,
    slot = "Seurat::BuildClusterTree"
  )
ape::plot.phylo(x = data.tree, direction = "downwards")
ape::nodelabels()
```

### [3.6.] Save & load data Seurat object 'pbmc_rna'
```{r eval=FALSE}
base::saveRDS(
    object = pbmc_rna,
    file = base::paste0(
        here::here("output_data/"),
        base::format(base::Sys.time(),"%Y%m%d"),
        "_pbmc_RNA",
        ".rds"
      ),
    compress = TRUE
  )
```

```{r eval=FALSE}
pbmc_rna <- base::list.files(
      path = here::here("output_data"),
      full.names = TRUE
    ) %>%
  stringr::str_subset("RNA.rds") %>%
  base::readRDS(
      file = .
    )
```

## [4.] Additional analysis & quality checks

### [4.0.] Subsample number of cells in initial scRNAseq features x cells matrix
#### 100,000 instead of 733,612 input cells (="base::ncol(pbmc_data$`Gene Expression`)").
```{r}
base::set.seed(1)
sub = 1e+5
cols <- base::sample(
    x = 1:base::ncol(pbmc_data$`Gene Expression`),
    size = sub,
    replace = FALSE
  )
pbmc_alt_data <- pbmc_data$`Gene Expression`[,cols]
```

### [4.1.] Run analysis with subsampled cells
#### No parameters changed comapred to [1.3.] - [3.4.] in '20241014_scRNAseq_analysis.Rmd'.
```{r}
base::set.seed(1)
pbmc_alt <- Seurat::CreateSeuratObject(
    counts = pbmc_alt_data,
    assay = "RNA",
    project = "pbmc10k",
    min.cells = 3,
    min.features = 200
  )
pbmc_alt[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_alt,
    pattern = "^MT-"
  )
pbmc_alt <- base::subset(
    x = pbmc_alt,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10
  )
pbmc_alt <- Seurat::NormalizeData(
    object = pbmc_alt,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
pbmc_alt <- Seurat::FindVariableFeatures(
    object = pbmc_alt,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
pbmc_alt <- Seurat::ScaleData(
    object = pbmc_alt,
    features = base::rownames(pbmc_alt),
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunPCA(
    object = pbmc_alt,
    features = Seurat::VariableFeatures(object = pbmc_alt),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
pbmc_alt <- Seurat::FindNeighbors(
    object = pbmc_alt,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
pbmc_alt <- Seurat::FindClusters(
    object = pbmc_alt,
    resolution = 0.5,
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunUMAP(
    object = pbmc_alt,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE
  )

cls <- pbmc_alt@meta.data %>%
  tibble::as_tibble(rownames = "cell") %>%
  dplyr::select(cell,seurat_clusters)
pbmc_alt@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell") %>% 
  dplyr::left_join(cls, by = "cell") %>% 
  ggplot2::ggplot(
      data = .,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        colour = seurat_clusters
      )
    ) +
  geom_point(
      size = 0.5
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      legend.position = "none"
    )

base::remove(pbmc_alt,pbmc_alt_data)
```

### [4.2.] Subsample reads in gene expression matrix
#### From 38.067.789 to 10.000.000 or 5.000.000.
#### See https://github.com/barkasn/nbHelpers/blob/master/R/subSampleCountMatrix.R
#### for 'nbHelpers::subSampleCountMatrix2()' function.
```{r}
base::set.seed(1)
fin = 5e6 # 1e+7
mat <- pbmc_data$`Gene Expression`
cur <- base::sum(mat@x)
rmv <- cur - fin
smp <- base::sample(
    x = base::sum(mat@x),
    size = rmv,
    replace = FALSE
  )
pos <- base::cumsum(mat@x)
t <- base::table(
    base::findInterval(
      x = smp,
      vec = pos,
      rightmost.closed = FALSE,
      left.open = TRUE
    ) + 1
  )
r <- base::as.numeric(t)
base::names(r) <- base::names(t)
t <- r
n <- t[base::match(base::as.character(1:base::length(mat@x)), base::names(t))]
n[base::is.na(n)] <- 0
mat@x <- mat@x - n # remove sampled reads
pbmc_alt_data <- Matrix::drop0(mat) # drop new 0s
base::sum(pbmc_alt_data@x)
base::remove(r,t,mat,cur,rmv,smp,pos,n)
```

### [4.3.] Run analysis with subsampled reads
#### No parameters changed comapred to [1.3.] - [3.4.] in '20241014_scRNAseq_analysis.Rmd'.
```{r}
base::set.seed(1)
pbmc_alt <- Seurat::CreateSeuratObject(
    counts = pbmc_alt_data,
    assay = "RNA",
    project = "reads10mio",
    min.cells = 3,
    min.features = 200
  )
pbmc_alt[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_alt,
    pattern = "^MT-"
  )
pbmc_alt <- base::subset(
    x = pbmc_alt,
    subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10
  )
pbmc_alt <- Seurat::NormalizeData(
    object = pbmc_alt,
    normalization.method = "LogNormalize",
    scale.factor = 1e+4,
    verbose = FALSE
  )
pbmc_alt <- Seurat::FindVariableFeatures(
    object = pbmc_alt,
    selection.method = "vst",
    nfeatures = 2000L,
    verbose = FALSE
  )
pbmc_alt <- Seurat::ScaleData(
    object = pbmc_alt,
    features = base::rownames(pbmc_alt),
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunPCA(
    object = pbmc_alt,
    features = Seurat::VariableFeatures(object = pbmc_alt),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "RNA", #default at this stage
    layer = 'scale.data' # default
  )
pbmc_alt <- Seurat::FindNeighbors(
    object = pbmc_alt,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
pbmc_alt <- Seurat::FindClusters(
    object = pbmc_alt,
    resolution = 0.5,
    verbose = FALSE
  )
pbmc_alt <- Seurat::RunUMAP(
    object = pbmc_alt,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE
  )

cls <- pbmc_alt@meta.data %>%
  tibble::as_tibble(rownames = "cell") %>%
  dplyr::select(cell,seurat_clusters)
pbmc_alt@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell") %>% 
  dplyr::left_join(cls, by = "cell") %>% 
  ggplot2::ggplot(
      data = .,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        colour = seurat_clusters
      )
    ) +
  geom_point(
      size = 0.1
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      legend.position = "none"
    )

# base::remove(pbmc_alt,pbmc_alt_data)
```

### [4.4.] Read out metainfo & UMAP data
```{r}
assertthat::are_equal(
    x = base::rownames(pbmc_rna@meta.data),
    y = base::rownames(pbmc_rna@reductions$umap@cell.embeddings)
  )

meta <- pbmc_rna@meta.data %>%
  tibble::as_tibble(rownames = "cell")
umap <- pbmc_rna@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell")

meta_comb <- dplyr::left_join(
    x = umap,
    y = meta,
    by = "cell"
  )
```

### [4.5.] Average mean cluster positions for cluster labels
```{r}
cls <- meta_comb %>%
  dplyr::group_by(seurat_clusters) %>%
  dplyr::summarise(
      pos_1 = base::mean(umap_1),
      pos_2 = base::mean(umap_2),
      .groups = "drop"
    )
```

### [4.6.] Map quality metrics onto UMAP
```{r}
coef = 0.5
cols <- base::c(
    grDevices::colorRampPalette(
        base::c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e")
      )(coef*100),
    grDevices::colorRampPalette(
      base::c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"),
      bias=2
    )((1-coef)*100)
  )

ggplot2::ggplot(
      data = meta_comb,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        color = nFeature_RNA # nFeature_RNA # percent.mt # nCount_RNA,
      )
    ) +
    geom_point(
      size = 0.1
    ) +
  ggrepel::geom_label_repel(
      inherit.aes = FALSE,
      data = cls,
      mapping = aes(
          x = pos_1,
          y = pos_2,
          label = seurat_clusters
        )
    ) +
  scale_colour_gradientn(
      colours = cols
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      axis.text = element_blank(),
      axis.ticks = element_blank()
    ) +
  guides(
    fill = "none",
    colour = guide_colourbar(
        label.position = "right",
        title = NULL,
        title.position = "left",
        title.theme = element_text(
            angle = 90,
            hjust = 0.5
          ),
        ticks = T,
        nbin = 100,
        barheight = 8,
        label = T,
        barwidth = 0.5
      )
    )
```

#### Please, insert & run your own code for Assignment question 10 here.
#### Make sure to save the resulting Seurat object as 'pbmc_qc' as shown in the next code chunk [4.7.].

### [4.7.] Save & load the "nFeature_RNA"-cleaned and annotated version pbmc_qc
```{r eval=FALSE}
base::saveRDS(
    object = pbmc_qc,
    file = base::paste0(
        here::here("output_data/"),
        base::format(base::Sys.time(),"%Y%m%d"),
        "_pbmc_RNA_qc_annotated",
        ".rds"
      ),
    compress = TRUE
  )
```

#### In case, problems occur with solving Assignment question 10 and no 'pbmc_qc' dataset
#### could be constructed, the next chunk will read in a backup 'pbmc_qc' dataset.

### [4.8.] Read in 'pbmc_qc' dataset
```{r}
pbmc_qc <- base::list.files(
      path = here::here("output_data"),
      full.names = TRUE
    ) %>%
  stringr::str_subset("pbmc_RNA_qc") %>%
  dplyr::last() %>% 
  base::readRDS(
      file = .
    )
# utils::str(pbmc_qc)
```

## [5.] Marker genes

### [5.0.] Identify all differential expressions
#### Check with 'base::range(pbmc_markers\$avg_log2FC)' and 'base::range(pbmc_markers\$p_val)'.
```{r}
pbmc_markers <- Seurat::FindAllMarkers(
    object = pbmc_qc,
    only.pos = TRUE,
    assay = "RNA",
    slot = "data", # "scale.data"
    test.use = 'wilcox',
    verbose = FALSE,
    random.seed = 1L,
    return.thresh = 1, # no limitation
    logfc.threshold = -Inf # no limitation
  )
```

### [5.1.] Inspect markers
#### Matching of canonical cell type markers to identified unbiased marker genes.
```{r}
pbmc_markers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(
      p_val_adj < 1e-2 & avg_log2FC > 1
    ) %>%
  dplyr::summarise(
      nFeat_sigDE = dplyr::n()
    )

pbmc_markers %>%
  dplyr::group_by(cluster) %>%
  dplyr::filter(
      p_val_adj < 1e-2
    ) %>%
  dplyr::arrange(cluster, dplyr::desc(avg_log2FC)) %>%
  dplyr::filter(
      dplyr::row_number() %in% 1:10
    ) %>%
  print(n = 100)
```

### [5.2.] Read out gene expression values from Seurat object
#### Default plotting via 'Seurat::FeaturePlot(pbmc_qc, features = "IL1A")'.
#### Access to genes via 'SeuratObject::Features(x = pbmc_qc,layer = "data")'.
```{r}
gene <- "IL1A" # "IL1B"

ge_comb <- Seurat::FetchData(
      object = pbmc_qc,
      vars = base::c(gene,"umap_1","umap_2"),
      cells = base::colnames(x = pbmc_qc),
      assay = "RNA",
      layer = "data"
    ) %>%
  dplyr::rename("gene" = 1)
```

### [5.3.] Plot gene expression values onto the UMAP projection
```{r}
coef = 0.5
cols <- base::c(
    grDevices::colorRampPalette(
        base::c("#e7f0fa", "#c9e2f6", "#95cbee", "#0099dc", "#4ab04a", "#ffd73e")
      )(coef*100),
    grDevices::colorRampPalette(
      base::c("#eec73a", "#e29421", "#e29421", "#f05336","#ce472e"),
      bias=2
    )((1-coef)*100)
  )

ggplot2::ggplot(
      data = ge_comb,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        color = gene
      )
    ) +
  geom_point(
      size = 0.1
    ) +
  scale_colour_gradientn(
      colours = cols
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      axis.text = element_blank(),
      axis.ticks = element_blank()
    )
```

### [5.4.] Read in reference data
```{r message=FALSE, warning=FALSE}
ref <- base::readRDS(
    file = base::paste0(
      here::here("input_data/"),
      "pbmc_10k_v3.rds"
    )
  )
ref <- Seurat::UpdateSeuratObject(object = ref)
```

### [5.5.] Inspect reference dataset
```{r}
cls <- ref@meta.data %>%
  tibble::as_tibble(rownames = "cell") %>%
  dplyr::select(cell,celltype)
ref@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell") %>% 
  dplyr::left_join(cls, by = "cell") %>% 
  ggplot2::ggplot(
      data = .,
      mapping = aes(
        x = UMAP_1,
        y = UMAP_2,
        colour = celltype
      )
    ) +
  geom_point(
      size = 0.5
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

### [5.6.] Transfer annotations via identified anchors
```{r}
pbmc_anchors <- Seurat::FindTransferAnchors(
    reference = ref,
    query = pbmc_qc,
    dims = 1:50,
    reference.reduction = "pca",
    verbose = FALSE
  )
predictions <- Seurat::TransferData(
    anchorset = pbmc_anchors,
    refdata = ref$celltype,
    dims = 1:50,
    verbose = FALSE
  )
pbmc_qc <- Seurat::AddMetaData(
    object = pbmc_qc,
    metadata = predictions
  )
```

### [5.7.] Label predicted IDs onto the UMAP dimension reduction
```{r}
pal <- base::c(
    "#3255A4", "#A4DBE8", "#8C8279",
    "#EA7600", "#F6BE00", "#28B463",
    "#AF7AC5", "#0E6655", "#0097A9",
    "#E03C31", "#B5BD00", "#500778",
    "#93272C", "#C6B0BC", "#8F993E",
    "#17202A", "#FF6F00", "#555025"
  )

meta_qc <- pbmc_qc@meta.data %>%
  tibble::as_tibble(rownames = "cell")
umap_qc <- pbmc_qc@reductions$umap@cell.embeddings %>%
  tibble::as_tibble(rownames = "cell")
meta_qc_comb <- dplyr::left_join(
    x = umap_qc,
    y = meta_qc,
    by = "cell"
  )

cls <- meta_qc_comb %>%
  dplyr::group_by(predicted.id) %>%
  dplyr::summarise(
      pos_1 = base::mean(umap_1),
      pos_2 = base::mean(umap_2),
      .groups = "drop"
    )

ggplot2::ggplot(
      data = meta_qc_comb,
      mapping = aes(
        x = umap_1,
        y = umap_2,
        color = predicted.id
      )
    ) +
  geom_point(
      size = 0.1
    ) +
  ggrepel::geom_label_repel(
      inherit.aes = FALSE,
      data = cls,
      mapping = aes(
          x = pos_1,
          y = pos_2,
          label = predicted.id
        ),
      box.padding = 1,
      max.overlaps = Inf
    ) +
  scale_colour_manual(
      values = pal
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1,
      axis.text = element_blank(),
      axis.ticks = element_blank()
    )
```

#### Extra question: Are platelets a real cluster or ambient RNA?

```{r eval=FALSE}
meta_qc_comb %>%
  dplyr::filter(predicted.id == "Platelet")
```

## [6.] Supplement / Appendix

### [6.0.] SCTransform
#### No QC filtering performed, which leads to inclusion of contaminations again.
```{r eval=FALSE}
base::set.seed(1L)
pbmc_sct <- Seurat::CreateSeuratObject(
    counts = pbmc_data$`Gene Expression`,
    assay = "RNA",
    project = "pbmcSCT",
    min.cells = 3,
    min.features = 200
  )
pbmc_sct[["percent.mt"]] <- Seurat::PercentageFeatureSet(
    object = pbmc_sct,
    pattern = "^MT-"
  )
# Necessary package: BiocManager::install('glmGamPoi')
pbmc_sct <- Seurat::SCTransform(
    object = pbmc_sct,
    verbose = FALSE,
    # vars.to.regress = "percent.mt"
  )
# utils::str(pbmc_sct)
pbmc_sct <- Seurat::RunPCA(
    object = pbmc_sct,
    features = Seurat::VariableFeatures(object = pbmc_sct),
    npcs = 50L,
    verbose = FALSE,
    seed.use = 1L,
    assay = "SCT",
    layer = 'scale.data' # default
  )
pbmc_sct <- Seurat::FindNeighbors(
    object = pbmc_sct,
    dims = 1:10,
    k.param = 20, # default
    verbose = FALSE,
    l2.norm = FALSE
  )
pbmc_sct <- Seurat::FindClusters(
    object = pbmc_sct,
    resolution = 0.3,
    verbose = FALSE
  )
pbmc_sct <- Seurat::RunUMAP(
    object = pbmc_sct,
    umap.method = "uwot",
    dims = 1:10,
    verbose = FALSE
  )
Seurat::DimPlot(
      object = pbmc_sct,
      reduction = "umap"
    ) +
  theme_bw() +
  theme(
      aspect.ratio = 1
    )
```

#### Extra question: Does SCTransform recapitulate the cell type distribution identified via transferred annotations better?
#### (e.g., better separation of small clusters like 'Double negative T-cells')

## [7.] Additional markdowns & vignettes for further analysis & reading

### [7.1.] Clustering

#### Significance hierarchical clustering - implemented as sc-SHC in R (ref.: https://www.nature.com/articles/s41592-023-01933-9]), see https://github.com/igrabski/sc-SHC
#### Scanorama implemented in Python with an R interface, see https://github.com/brianhie/scanorama/blob/master/bin/R/scanorama.R

### [7.2.] Trajectory analysis

#### Use Monocle3 function wrapped in Seurat pipeline to perform trajectory analysis, see https://htmlpreview.github.io/?https://github.com/satijalab/seurat-wrappers/blob/master/docs/monocle3.html
#### and also https://rpubs.com/mahima_bose/Seurat_and_Monocle3_p
#### Trajectory analysis via Diffusion maps, Diffusion pseudotime analysis or slingshot, see https://broadinstitute.github.io/2019_scWorkshop/functional-pseudotime-analysis.html
